// uu_appg01_workspace/src/api/errors/sys-app-secret-store-errors.js
const WorkspaceError = require("./workspace-error.js");

const PutSecret = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_LIB_ERROR_PREFIX}sys/uuAppSecretStore/putSecret/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${PutSecret.UC_CODE}invalidDtoIn`;
    }
  },

  SysUuAppSecretStoreDaoUpdateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Updating of sysUuAppSecret failed.";
      this.code = `${PutSecret.UC_CODE}updateFailed`;
    }
  },

  SysUuAppSecretStoreDaoCreateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Creation of sysUuAppSecret failed.";
      this.code = `${PutSecret.UC_CODE}createFailed`;
    }
  },
};

const GetSecret = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_LIB_ERROR_PREFIX}sys/uuAppSecretStore/getSecret/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${GetSecret.UC_CODE}invalidDtoIn`;
    }
  },

  MasterKeyNotFound: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Unable to decrypt secret because master keys are not configured for the slot that was used for encryption.";
      this.code = `${GetSecret.UC_CODE}masterKeyNotFound`;
    }
  },
};

const RemoveSecret = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_LIB_ERROR_PREFIX}sys/uuAppSecretStore/removeSecret/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${RemoveSecret.UC_CODE}invalidDtoIn`;
    }
  },

  SysUuAppSecretStoreDaoDeleteFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Removing of sysUuAppSecret failed.";
      this.code = `${RemoveSecret.UC_CODE}deleteFailed`;
    }
  },
};

const ListCodes = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_LIB_ERROR_PREFIX}sys/uuAppSecretStore/listSecretCodes/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${ListCodes.UC_CODE}invalidDtoIn`;
    }
  },
};

const RotateMasterKey = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_LIB_ERROR_PREFIX}sys/uuAppSecretStore/rotateMasterKey/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${RotateMasterKey.UC_CODE}invalidDtoIn`;
    }
  },

  SysUuSubAppInstanceDoesNotExist: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysUuSubAppInstance does not exist.";
      this.code = `${RotateMasterKey.UC_CODE}sysUuSubAppInstanceDoesNotExist`;
    }
  },

  SysUuSubAppInstanceDaoUpdateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to update sysUuSubAppInstance object.";
      this.code = `${RotateMasterKey.UC_CODE}sysUuSubAppInstanceDaoUpdateFailed`;
    }
  },

};

module.exports = {
  PutSecret,
  GetSecret,
  RemoveSecret,
  ListCodes,
  RotateMasterKey,
};
