// uu_appg01_workspace/src/api/errors/sys-app-workspace-errors.js
const WorkspaceError = require("./workspace-error.js");
const {BaseError} = require("uu_appg01_core-utils");

const Create = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}create/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${Create.UC_CODE}invalidDtoIn`;
    }
  },
  UserIsNotAuthorized: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.status = 403;
      this.message = "Logged user is not authorized to run this use case.";
      this.code = `${Create.UC_CODE}userIsNotAuthorized`;
    }
  },
  ApplicationSignatureCheckFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.status = 403;
      this.message = "uuAppClientToken is invalid.";
      this.code = `${Create.UC_CODE}applicationSignatureCheckFailed`;
    }
  },
  SysUuAppWorkspaceAlreadyCreated: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysUuAppWorkspace has already been created.";
      this.code = `${Create.UC_CODE}sysUuAppWorkspaceAlreadyCreated`;
    }
  },

  AwidConflictsWithAsid: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysUuAppWorkspace cannot be initialized. Awid must not be the same as asid.";
      this.code = `${Create.UC_CODE}awidConflictsWithAsid`;
    }
  },

  SysUuAppWorkspaceCreationInProgress: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysUuAppWorkspace is currently beeing created.";
      this.code = `${Create.UC_CODE}sysUuAppWorkspaceCreationInProgress`;
    }
  },

  SysUuSubAppInstanceDoesNotExist: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "AppInstance does not exist.";
      this.code = `${Create.UC_CODE}sysAppInstanceDoesNotExist`;
    }
  },

  CreateSchemaFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Creation of uuObject schema by uuObject DAO createSchema failed.";
      this.code = `${Create.UC_CODE}uuObjectDaoCreateSchemaFailed`;
    }
  },

  SysUuAppWorkspaceDaoCreateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Creation of SysUuAppWorkspace by SysUuAppWorkspace DAO create failed.";
      this.code = `${Create.UC_CODE}sysUuAppWorkspaceDaoCreateFailed`;
    }
  },

  SysProfileDaoCreateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Creation of sysProfile by DAO create failed.";
      this.code = `${Create.UC_CODE}sysProfileDaoCreateFailed`;
    }
  },

  SysUuAppWorkspaceConfigDaoCreateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Creation of SysUuAppWorkspaceConfig by DAO create failed.";
      this.code = `${Create.UC_CODE}sysUuAppWorkspaceConfigDaoCreateFailed`;
    }
  },

  SysProfileDaoDeleteByAwidFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Delete of sysProfile by sysProfile Dao deleteByAwid failed.";
      this.code = `${Create.UC_CODE}sysProfileDaoDeleteByAwidFailed`;
    }
  },

  SysUuAppWorkspaceConfigDaoDeleteByAwidFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Delete of SysUuAppWorkspaceConfig by SysUuAppWorkspaceConfig Dao delete failed.";
      this.code = `${Create.UC_CODE}sysUuAppWorkspaceConfigDaoDeleteByAwidFailed`;
    }
  },

  SysUuAppWorkspaceDaoDeleteByAwidFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Delete of SysUuAppWorkspace by SysUuAppWorkspace Dao delete failed.";
      this.code = `${Create.UC_CODE}sysUuAppWorkspaceDaoDeleteByAwidFailed`;
    }
  },

  SysUuAppWorkspaceDaoUpdateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Update of SysUuAppWorkspace by SysUuAppWorkspace Dao updateByAwid failed.";
      this.code = `${Create.UC_CODE}sysUuAppWorkspaceDaoUpdateFailed`;
    }
  },

  SysUuAppWorkspaceDataStoreStatsDaoCreateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Creation of sysDataStoreStats by DAO create failed.";
      this.code = `${Create.UC_CODE}sysUuAppWorkspaceDataStoreStatsDaoCreateFailedError`;
    }
  }
};

const Get = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}get/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${Get.UC_CODE}invalidDtoIn`;
    }
  },

  SysUuAppWorkspaceDoesNotExistError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Object sysUuAppWorkspace does not exist.";
      this.code = `${Get.UC_CODE}sysUuAppWorkspaceDoesNotExist`;
    }
  }
};

const Set = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}set/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${Set.UC_CODE}invalidDtoIn`;
    }
  },

  SysUuAppWorkspaceDoesNotExistError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysUuAppWorkspace does not exist.";
      this.code = `${Set.UC_CODE}sysUuAppWorkspaceDoesNotExist`;
    }
  },

  SysUuAppWorkspaceDaoUpdateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to update SysUuAppWorkspace object.";
      this.code = `${Set.UC_CODE}sysUuAppWorkspaceDaoUpdateFailed`;
    }
  }
};

const SetActiveSysState = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}setActiveSysState/`,

  InvalidDtoIn: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SetActiveSysState.UC_CODE}invalidDtoIn`;
    }
  },

  SysUuAppWorkspaceDoesNotExist: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysUuAppWorkspace does not exist.";
      this.code = `${SetActiveSysState.UC_CODE}sysUuAppWorkspaceDoesNotExist`;
    }
  },

  SysUuAppWorkspaceDaoUpdateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to set new sysState using SysUuAppWorkspace DAO update.";
      this.code = `${SetActiveSysState.UC_CODE}sysUuAppWorkspaceDaoUpdateFailed`;
    }
  }
};

const SetRestrictedSysState = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}setRestrictedSysState/`,

  InvalidDtoIn: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SetRestrictedSysState.UC_CODE}invalidDtoIn`;
    }
  },

  SysUuAppWorkspaceDoesNotExist: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysUuAppWorkspace does not exist.";
      this.code = `${SetRestrictedSysState.UC_CODE}sysUuAppWorkspaceDoesNotExist`;
    }
  },

  SysUuAppWorkspaceDaoUpdateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to set new sysState using SysUuAppWorkspace DAO update.";
      this.code = `${SetRestrictedSysState.UC_CODE}sysUuAppWorkspaceDaoUpdateFailed`;
    }
  }
};

const SetReadOnlySysState = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}setReadOnlySysState/`,

  InvalidDtoIn: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SetReadOnlySysState.UC_CODE}invalidDtoIn`;
    }
  },

  SysUuAppWorkspaceDoesNotExist: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysUuAppWorkspace does not exist.";
      this.code = `${SetReadOnlySysState.UC_CODE}sysUuAppWorkspaceDoesNotExist`;
    }
  },

  SysUuAppWorkspaceDaoUpdateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to set new sysState using SysUuAppWorkspace DAO update.";
      this.code = `${SetReadOnlySysState.UC_CODE}sysUuAppWorkspaceDaoUpdateFailed`;
    }
  }
};

const SetAuthorizationStrategy = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}setAuthorizationStrategy/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SetAuthorizationStrategy.UC_CODE}invalidDtoIn`;
    }
  },
  UuAppWorkspaceIsNotInProperStateError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysUuAppWorkspace is not in proper state.";
      this.code = `${SetAuthorizationStrategy.UC_CODE}uuAppWorkspaceIsNotInProperState`;
    }
  },
  UserIsNotAuthorizedToChangeAuthStrategy: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Logged user is not authorized to change authorization strategy.";
      this.code = `${SetAuthorizationStrategy.UC_CODE}userIsNotAuthorizedToChangeAuthStrategy`;
    }
  },
  SysUuAppWorkspaceDoesNotExist: class extends  WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysUuAppWorkspace does not exist.";
      this.code = `${SetAuthorizationStrategy.UC_CODE}sysUuAppWorkspaceDoesNotExist`;
    }
  },
  SysUuAppWorkspaceDaoUpdateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Update of SysUuAppWorkspace by SysUuAppWorkspace Dao updateByAwid failed.";
      this.code = `${SetAuthorizationStrategy.UC_CODE}sysUuAppWorkspaceDaoUpdateFailed`;
    }
  }
};

const ProfileGet = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}profile/get/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${ProfileGet.UC_CODE}invalidDtoIn`;
    }
  },

  SysProfileDoesNotExistError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysProfile does not exist.";
      this.code = `${ProfileGet.UC_CODE}sysProfileDoesNotExist`;
    }
  }
};

const ProfileSet = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}profile/set/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${ProfileSet.UC_CODE}invalidDtoIn`;
    }
  },

  SysProfileDoesNotExistError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysProfile does not exist.";
      this.code = `${ProfileSet.UC_CODE}sysProfileDoesNotExist`;
    }
  },

  SysProfileDaoUpdateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to update sysProfile (DAO update failed).";
      this.code = `${ProfileSet.UC_CODE}sysProfileDaoUpdateFailed`;
    }
  }
};

const ProfileList = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}profile/list/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${ProfileList.UC_CODE}invalidDtoIn`;
    }
  }

};

const ProfileUpdate = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}profile/update/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${ProfileUpdate.UC_CODE}invalidDtoIn`;
    }
  },
  NoAwidInProperState: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "There is no awid in proper state to be initialized.";
      this.code = `${ProfileUpdate.UC_CODE}noAwidInProperState`;
    }
  },
  SysUuAppWorkspaceDoesNotExistError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysUuAppWorkspace does not exist or is not in initialized state.";
      this.code = `${ProfileUpdate.UC_CODE}sysUuAppWorkspaceDoesNotExist`;
    }
  },
  SysUuAppWorkspaceDaoUpdateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Update of SysUuAppWorkspace by SysUuAppWorkspace Dao updateByAwid failed.";
      this.code = `${ProfileUpdate.UC_CODE}sysUuAppWorkspaceDaoUpdateFailed`;
    }
  },
  SysProfileDaoCreateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to create a new sysProfile.";
      this.code = `${ProfileUpdate.UC_CODE}sysProfileDaoCreateFailed`;
    }
  },
  SysProfileDaoDeleteFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to delete a sysProfile.";
      this.code = `${ProfileUpdate.UC_CODE}sysProfileDaoDeleteFailed`;
    }
  }
};

const ConnectArtifact = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}connectArtifact/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${ConnectArtifact.UC_CODE}invalidDtoIn`;
    }
  },
  UuAppWorkspaceIsNotInProperStateError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysUuAppWorkspace is not in proper state.";
      this.code = `${ConnectArtifact.UC_CODE}uuAppWorkspaceIsNotInProperState`;
    }
  },
  UuCmdConnectAppWorkspaceCallFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "uuCMD uuAwsc/connectAppWorkspace could not be started.";
      this.code = `${ConnectArtifact.UC_CODE}uuCmdConnectAppWorkspaceCallFailed`;
    }
  },
  SysUuAppWorkspaceDaoUpdateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Update of SysUuAppWorkspace by SysUuAppWorkspace Dao updateByAwid failed.";
      this.code = `${ConnectArtifact.UC_CODE}sysUuAppWorkspaceDaoUpdateFailed`;
    }
  },
  TokenCreateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message =
        "Failed to create uuAppClientToken because keys for it's signature were not initialized. Run the uuCmd sys/appClientToken/initKeys first.";
      this.code = `${ConnectArtifact.UC_CODE}uuAppClientTokenCreateFailed`;
    }
  }
};

const Delete = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}delete/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${Delete.UC_CODE}invalidDtoIn`;
    }
  },
  SysUuAppWorkspaceNotInProperState: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysUuAppWorkspace is not in proper state.";
      this.code = `${Delete.UC_CODE}sysUuAppWorkspaceNotInProperState`;
    }
  },
  SysUuAppWorkspaceDoesNotExist: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysUuAppWorkspace does not exist.";
      this.code = `${Delete.UC_CODE}sysUuAppWorkspaceDoesNotExist`;
    }
  },
  UuAwscConnectedToTheSysUuAppWorkspace: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "There is uuAwsc connected to the SysUuAppWorkspace.";
      this.code = `${Delete.UC_CODE}uuAwscConnectedToTheSysUuAppWorkspace`;
    }
  }
};

const PermissionDelete = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}permission/delete/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${PermissionDelete.UC_CODE}invalidDtoIn`;
    }
  },

  SysPermissionDeleteDaoFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to delete given SysPermission.";
      this.code = `${PermissionDelete.UC_CODE}sysPermissionDeleteDaoFailed`;
    }
  }
};

const PermissionList = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}permission/list/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${PermissionList.UC_CODE}invalidDtoIn`;
    }
  }
};

const PermissionCreate = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}permission/create/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${PermissionCreate.UC_CODE}invalidDtoIn`;
    }
  },

  SysProfileDoesNotExistError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "The given profile does not exist.";
      this.code = `${PermissionCreate.UC_CODE}sysProfileDoesNotExist`;
    }
  },

  SysPermissionMaxNumberExceeded: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message =
        "Max number of sysPermission uuObjects would be exceeded.";
      this.code = `${PermissionCreate.UC_CODE}sysPermissionMaxNumberExceeded`;
    }
  },

  SysPermissionDaoCreateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to create a new sysPermission uuObject.";
      this.code = `${PermissionCreate.UC_CODE}sysPermissionDaoCreateFailed`;
    }
  }
};

const Close = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}close/`,

  InvalidDtoIn: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${Close.UC_CODE}invalidDtoIn`;
    }
  },

  SysUuAppWorkspaceDoesNotExist: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysUuAppWorkspace with given awid does not exist.";
      this.code = `${Close.UC_CODE}sysUuAppWorkspaceDoesNotExist`;
    }
  },

  SetStateInUuBaseRegistryFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to set uuAppWorkspace state in uuBaseRegistry.";
      this.code = `${Close.UC_CODE}setStateInUuBaseRegistryFailed`;
    }
  },

  SysUuAppWorkspaceAlreadyClosed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "uuAppWorkspace is already closed.";
      this.code = `${Close.UC_CODE}sysUuAppWorkspaceAlreadyClosed`;
    }
  },

  SysUuAppWorkspaceDaoUpdateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to update uuAppWorkspace state.";
      this.code = `${Close.UC_CODE}sysUuAppWorkspaceDaoUpdateFailed`;
    }
  }
};

const Suspend = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}suspend/`,

  InvalidDtoIn: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${Suspend.UC_CODE}invalidDtoIn`;
    }
  },

  SysUuAppWorkspaceDoesNotExist: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysUuAppWorkspace with given awid does not exist.";
      this.code = `${Suspend.UC_CODE}sysUuAppWorkspaceDoesNotExist`;
    }
  },

  SetStateInUuBaseRegistryFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to set uuAppWorkspace state in uuBaseRegistry.";
      this.code = `${Suspend.UC_CODE}setStateInUuBaseRegistryFailed`;
    }
  },

  SysUuAppWorkspaceNotInProperState: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysUuAppWorkspace is not in proper state.";
      this.code = `${Suspend.UC_CODE}sysUuAppWorkspaceNotInProperState`;
    }
  },

  SysUuAppWorkspaceDaoUpdateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to update uuAppWorkspace state.";
      this.code = `${Suspend.UC_CODE}sysUuAppWorkspaceDaoUpdateFailed`;
    }
  }
};

const Resume = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}resume/`,

  InvalidDtoIn: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${Resume.UC_CODE}invalidDtoIn`;
    }
  },

  SysUuAppWorkspaceDoesNotExist: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysUuAppWorkspace with given awid does not exist.";
      this.code = `${Resume.UC_CODE}sysUuAppWorkspaceDoesNotExist`;
    }
  },

  SetStateInUuBaseRegistryFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to set uuAppWorkspace state in uuBaseRegistry.";
      this.code = `${Resume.UC_CODE}setStateInUuBaseRegistryFailed`;
    }
  },

  sysUuAppWorkspaceNotInSuspendedState: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "uuAppWorkspace is not in suspended state.";
      this.code = `${Resume.UC_CODE}sysUuAppWorkspaceNotInProperState`;
    }
  },

  SysUuAppWorkspaceDaoUpdateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to update uuAppWorkspace state.";
      this.code = `${Resume.UC_CODE}sysUuAppWorkspaceDaoUpdateFailed`;
    }
  }
};

const List = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}list/`,

  InvalidDtoIn: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${List.UC_CODE}invalidDtoIn`;
    }
  }
};

const SetConfigParameter = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}setConfigParameter/`,

  InvalidDtoIn: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SetConfigParameter.UC_CODE}invalidDtoIn`;
    }
  },

  SysUuAppWorkspaceDoesNotExist: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysUuAppWorkspace does not exist.";
      this.code = `${SetConfigParameter.UC_CODE}sysUuAppWorkspaceDoesNotExist`;
    }
  },

  SysUuAppWorkspaceDaoUpdateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to set restricted sysState using sysAppInstance DAO update.";
      this.code = `${SetConfigParameter.UC_CODE}sysUuAppWorkspaceDaoUpdateFailed`;
    }
  }

};

const GetConfigParameter = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}getConfigParameter/`,

  InvalidDtoIn: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${GetConfigParameter.UC_CODE}invalidDtoIn`;
    }
  }

};

class MaximumNumberAwidsError extends BaseError {
  constructor(maxNoa) {
    super(`Inserting uuObject failed. Maximum number of awids (${maxNoa}) reached.`);
  }
}

module.exports = {
  Create,
  Get,
  Set,
  MaximumNumberAwidsError,
  SetAuthorizationStrategy,
  ConnectArtifact,
  Delete,
  ProfileGet,
  ProfileSet,
  ProfileList,
  ProfileUpdate,
  PermissionDelete,
  PermissionList,
  PermissionCreate,
  Close,
  Suspend,
  Resume,
  SetActiveSysState,
  SetRestrictedSysState,
  SetReadOnlySysState,
  List,
  SetConfigParameter,
  GetConfigParameter
};
