// uu_appg01_workspace/src/api/errors/sys-app-instance-errors.js
const WorkspaceError = require("./workspace-error.js");

const Init = {
  UC_CODE: `${WorkspaceError.APP_INSTANCE_ERROR_PREFIX}init/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${Init.UC_CODE}invalidDtoIn`;
    }
  },
  SysUuSubAppInstanceIsAlreadyInitialized: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "The uuSubAppInstance has already been initialized.";
      this.code = `${Init.UC_CODE}sysUuSubAppInstanceIsAlreadyInitialized`;
    }
  },
  SysUuSubAppInstanceDaoCreateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to create sysUuSubAppInstance object.";
      this.code = `${Init.UC_CODE}sysUuSubAppInstanceDaoCreateFailed`;
    }
  },
  BaseRegistryUriVerificationFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Verification of baseRegistryUri failed.";
      this.code = `${Init.UC_CODE}baseRegistryUriVerificationFailed`;
    }
  },
  UserIsNotAuthorized: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.status = 403;
      this.message = "Logged user is not authorized to run the use case.";
      this.code = `${Init.UC_CODE}userIsNotAuthorized`;
    }
  },
  ApplicationIsNotAuthorized: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.status = 403;
      this.message = "Application is not authorized to run the use case.";
      this.code = `${Init.UC_CODE}applicationIsNotAuthorized`;
    }
  },
  ApplicationSignatureCheckFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to validate uuAppClientToken.";
      this.code = `${Init.UC_CODE}applicationSignatureCheckFailed`;
    }
  },
  SysUuSubAppInstanceAlreadyStartedAndLocked: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "The uuSubAppInstance initialization is in process.";
      this.code = `${Init.UC_CODE}sysAppInstanceAlreadyStartedAndLocked`;
    }
  },
  SysUuSubAppInstanceDaoUpdateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to update state of the uuSubAppInstance.";
      this.code = `${Init.UC_CODE}sysUuSubAppInstanceDaoUpdateFailed`;
    }
  },
  CreateSchemaDaoFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to create a schema.";
      this.code = `${Init.UC_CODE}createSchemaDaoFailed`;
    }
  },
  SysProfileDaoCreateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to create sysProfile object.";
      this.code = `${Init.UC_CODE}sysProfileDaoCreateFailed`;
    }
  },
  SysProfileDaoUpdateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to set roleGroupUri of AsidAuthorities sysProfile.";
      this.code = `${Init.UC_CODE}sysProfileDaoUpdateFailed`;
    }
  },
  SysUuAppInstanceDataStoreStatsDaoCreateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Creation of sysDataStoreStats by DAO create failed.";
      this.code = `${Init.UC_CODE}sysUuSubAppInstanceDataStoreStatsDaoCreateFailed`;
    }
  }
};

const Get = {
  UC_CODE: `${WorkspaceError.APP_INSTANCE_ERROR_PREFIX}get/`,

  SysUuSubAppInstanceDoesNotExistError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Object sysUuSubAppInstance does not exist.";
      this.code = `${Get.UC_CODE}sysUuSubAppInstanceDoesNotExist`;
    }
  }
};

const Set = {
  UC_CODE: `${WorkspaceError.APP_INSTANCE_ERROR_PREFIX}set/`,

  InvalidDtoIn: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${Set.UC_CODE}invalidDtoIn`;
    }
  },

  SysUuSubAppInstanceDoesNotExist: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "sysUuSubAppInstance object does not exist.";
      this.code = `${Set.UC_CODE}sysUuSubAppInstanceDoesNotExist`;
    }
  },

  SysUuSubAppInstanceDaoUpdateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to update sysUuSubAppInstance object.";
      this.code = `${Init.UC_CODE}sysUuSubAppInstanceDaoUpdateFailed`;
    }
  }
};

const ProfileGet = {
  UC_CODE: `${WorkspaceError.APP_INSTANCE_ERROR_PREFIX}profile/get/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${ProfileGet.UC_CODE}invalidDtoIn`;
    }
  },

  SysProfileDoesNotExistError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysProfile does not exist.";
      this.code = `${ProfileGet.UC_CODE}sysProfileDoesNotExist`;
    }
  }
};

const ProfileSet = {
  UC_CODE: `${WorkspaceError.APP_INSTANCE_ERROR_PREFIX}profile/set/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${ProfileSet.UC_CODE}invalidDtoIn`;
    }
  },

  SysProfileDoesNotExistError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysProfile does not exist.";
      this.code = `${ProfileSet.UC_CODE}sysProfileDoesNotExist`;
    }
  },

  SysProfileDaoUpdateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to update sysProfile (DAO update failed).";
      this.code = `${ProfileSet.UC_CODE}sysProfileDaoUpdateFailed`;
    }
  }
};

const ProfileList = {
  UC_CODE: `${WorkspaceError.APP_INSTANCE_ERROR_PREFIX}profile/list/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${ProfileList.UC_CODE}invalidDtoIn`;
    }
  }

};

const PermissionDelete = {
  UC_CODE: `${WorkspaceError.APP_INSTANCE_ERROR_PREFIX}permission/delete/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${PermissionDelete.UC_CODE}invalidDtoIn`;
    }
  },

  SysPermissionDeleteDaoFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to delete given SysPermission.";
      this.code = `${PermissionDelete.UC_CODE}sysPermissionDeleteDaoFailed`;
    }
  }
};

const PermissionList = {
  UC_CODE: `${WorkspaceError.APP_INSTANCE_ERROR_PREFIX}permission/list/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${PermissionList.UC_CODE}invalidDtoIn`;
    }
  }
};

const PermissionCreate = {
  UC_CODE: `${WorkspaceError.APP_INSTANCE_ERROR_PREFIX}permission/create/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${PermissionCreate.UC_CODE}invalidDtoIn`;
    }
  },

  SysProfileDoesNotExistError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "The given profile does not exist.";
      this.code = `${PermissionCreate.UC_CODE}sysProfileDoesNotExist`;
    }
  },

  SysPermissionMaxNumberExceeded: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message =
        "Max number of sysPermission uuObjects would be exceeded.";
      this.code = `${PermissionCreate.UC_CODE}sysPermissionMaxNumberExceeded`;
    }
  },

  SysPermissionDaoCreateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to create a new sysPermission uuObject.";
      this.code = `${PermissionCreate.UC_CODE}sysPermissionDaoCreateFailed`;
    }
  }
};

const SetActiveSysState = {
  UC_CODE: `${WorkspaceError.APP_INSTANCE_ERROR_PREFIX}setActiveSysState/`,

  InvalidDtoIn: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SetActiveSysState.UC_CODE}invalidDtoIn`;
    }
  },

  SysUuSubAppInstanceDoesNotExist: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysUuSubAppInstance does not exist.";
      this.code = `${SetActiveSysState.UC_CODE}sysUuSubAppInstanceDoesNotExist`;
    }
  },

  SysUuSubAppInstanceDaoUpdateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to set active sysState using sysUuSubAppInstance DAO update.";
      this.code = `${SetActiveSysState.UC_CODE}sysUuSubAppInstanceDaoUpdateFailed`;
    }
  }
};

const SetRestrictedSysState = {
  UC_CODE: `${WorkspaceError.APP_INSTANCE_ERROR_PREFIX}setRestrictedSysState/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SetRestrictedSysState.UC_CODE}invalidDtoIn`;
    }
  },

  SysUuSubAppInstanceDoesNotExist: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysUuSubAppInstance does not exist.";
      this.code = `${SetRestrictedSysState.UC_CODE}sysUuSubAppInstanceDoesNotExist`;
    }
  },

  SysUuSubAppInstanceDaoUpdateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to set restricted sysState using sysUuSubAppInstance DAO update.";
      this.code = `${SetRestrictedSysState.UC_CODE}sysUuSubAppInstanceDaoUpdateFailed`;
    }
  }
};

const SetConfigParameter = {
  UC_CODE: `${WorkspaceError.APP_INSTANCE_ERROR_PREFIX}setConfigParameter/`,

  InvalidDtoIn: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SetConfigParameter.UC_CODE}invalidDtoIn`;
    }
  },

  SysUuSubAppInstanceDoesNotExist: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysUuSubAppInstance does not exist.";
      this.code = `${SetConfigParameter.UC_CODE}sysUuSubAppInstanceDoesNotExist`;
    }
  },

  SysUuSubAppInstanceDaoUpdateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Failed to set restricted sysState using sysUuSubAppInstance DAO update.";
      this.code = `${SetConfigParameter.UC_CODE}sysUuSubAppInstanceDaoUpdateFailed`;
    }
  }

};

const GetConfigParameter = {
  UC_CODE: `${WorkspaceError.APP_INSTANCE_ERROR_PREFIX}getConfigParameter/`,

  InvalidDtoIn: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${GetConfigParameter.UC_CODE}invalidDtoIn`;
    }
  }

};


module.exports = {
  Init,
  Get,
  Set,
  ProfileGet,
  ProfileSet,
  ProfileList,
  PermissionDelete,
  PermissionList,
  PermissionCreate,
  SetActiveSysState,
  SetRestrictedSysState,
  SetConfigParameter,
  GetConfigParameter
};
