// uu_appg01_workspace/src/api/errors/sys-dump-restore-errors.js
const WorkspaceError = require("./workspace-error.js");

const SysDump = {
  UC_CODE: `${WorkspaceError.APP_INSTANCE_ERROR_PREFIX}dump/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SysDump.UC_CODE}invalidDtoIn`;
    }
  },
  MaximumSizeExceeded: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Dump size exceeds the maximum allowed size.";
      this.code = `${SysDump.UC_CODE}maximumSizeExceeded`;
    }
  },
  MaximumSizeExceededNoStats: class extends WorkspaceError {
    //update binary data
    constructor() {
      super(...arguments);
      this.message =
        "Estimated dump size exceeds the maximum allowed size. Estimated dump size may be inaccurate" +
        " because it was not obtained based on the uuAppDataStore statistics as it has not yet been calculated." +
        " Please run uuCmd sys/calculateDataStoreStats and try to run dump again.";
      this.code = `${SysDump.UC_CODE}maximumSizeExceededNoStats`;
    }
  }
};

const SysRestore = {
  UC_CODE: `${WorkspaceError.APP_INSTANCE_ERROR_PREFIX}restore/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SysRestore.UC_CODE}invalidDtoIn`;
    }
  },
  MetadataNotFound: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "File _metadata.json not found in input data.";
      this.code = `${SysRestore.UC_CODE}metadataNotFound`;
    }
  },
  InvalidMetadata: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Metadata is not valid.";
      this.code = `${SysRestore.UC_CODE}invalidMetadata`;
    }
  },
  InvalidSchemaMetadata: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Metadata is not valid.";
      this.code = `${SysRestore.UC_CODE}invalidSchemaMetadata`;
    }
  },
  DaoNotFound: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Dao not found for schema from input data.";
      this.code = `${SysRestore.UC_CODE}daoNotFound`;
    }
  },
  InvalidDump: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Dump is not valid. See the cause for more details.";
      this.code = `${SysRestore.UC_CODE}invalidDump`;
    }
  },
  artifactConnectedToRestoredAwid: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Awid can not be restored because it already exists and uuBT artifact is connected to it.";
      this.code = `${SysRestore.UC_CODE}artifactConnectedToRestoredAwid`;
    }
  },
  AwidNotFoundInDump: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Dump does not contain requested AWID.";
      this.code = `${SysRestore.UC_CODE}awidNotFoundInDump`;
    }
  },
  RestoreFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Restore by DAO failed. See the cause for more details.";
      this.code = `${SysRestore.UC_CODE}restoreFailed`;
    }
  },
  AsidNotFoundInDump: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Dump does not contain ASID of your application.";
      this.code = `${SysRestore.UC_CODE}asidNotFoundInDump`;
    }
  },
  InvalidDumpRow: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Dump row is not valid. See the cause for more details.";
      this.code = `${SysRestore.UC_CODE}invalidDumpRow`;
    }
  }
};

const SysAwidRestore = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}restore/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SysAwidRestore.UC_CODE}invalidDtoIn`;
    }
  },
  InvalidDumpRow: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Dump row is not valid. See the cause for more details.";
      this.code = `${SysAwidRestore.UC_CODE}invalidDumpRow`;
    }
  }
};

const SysImport = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}import/`,
  SysUuAppWorkspaceDoesNotExist: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Object sysUuAppWorkspace does not exist.";
      this.code = `${SysImport.UC_CODE}sysUuAppWorkspaceDoesNotExist`;
    }
  },
  InvalidDtoIn: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SysImport.UC_CODE}invalidDtoIn`;
    }
  },
  InvalidSchemaMetadata: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Metadata is not valid.";
      this.code = `${SysImport.UC_CODE}invalidSchemaMetadata`;
    }
  },
  DaoNotFound: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Dao not found for schema from input data.";
      this.code = `${SysImport.UC_CODE}daoNotFound`;
    }
  },
  InvalidDump: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Dump is not valid.";
      this.code = `${SysImport.UC_CODE}invalidDump`;
    }
  },
  ImportFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Import failed. See the cause for more details.";
      this.code = `${SysImport.UC_CODE}importFailed`;
    }
  },
  SysImportMappingDaoCreateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Creating record by SysImportMapping DAO create failed.";
      this.code = `${SysImport.UC_CODE}sysImportMappingDaoCreateFailed`;
    }
  }
};

const SysListImportMappings = {
  UC_CODE: `${WorkspaceError.APP_WORKSPACE_ERROR_PREFIX}listImportMappings/`,
  InvalidDtoIn: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SysListImportMappings.UC_CODE}invalidDtoIn`;
    }
  }
};


module.exports = {
  SysDump,
  SysRestore,
  SysAwidRestore,
  SysImport,
  SysListImportMappings
};
