// uu_appg01_workspace/src/dao/sys-import-mapping-mongo.js
"use strict";

const { UuObjectDao } = require("uu_appg01_objectstore");
const { ObjectId } = require("bson");

class SysImportMappingMongo extends UuObjectDao {

  async createSchema() {
    await super.createIndex(
      {
        awid: 1,
        origId: 1
      });
  }

  async create(recordList) {
    let result = await super.insertMany(recordList);
    if (!result.succesful) {
      throw result.error;
    }
    return result;
  }

  async delete(awid) {
    await super.deleteMany({ awid });
  }

  async list(awid, pageInfo) {
    let sort = { "sys.cts": -1 };
    return await super.find({ awid }, pageInfo, sort);
  }

  async listByOrigIdList(awid, origIdList, pageInfo) {
    let sort = { "sys.cts": -1 };
    let query = {
      awid,
      origId: {
        "$in": origIdList.map(x => {
          return ObjectId.isValid(x) ? new ObjectId(x) : x;
        })
      }
    };
    return await super.find(query, pageInfo, sort);
  }

  getObjectId() {
    return new ObjectId();
  }

}

module.exports = SysImportMappingMongo;
