// uu_appg01_server/src/index.js
"use strict";

module.exports = {
  get start() {
    return require("uu_appg01_core-appserver").startServer;
  },
  get stop() {
    return require("uu_appg01_core-appserver").stopServer;
  },
  get disableMiddleware() {
    return require("uu_appg01_core-appserver").disableMiddleware;
  },
  get AppClient() {
    const { AppClient, RemoteError, ApplicationError, SystemError } = require("uu_appg01_server-client");
    AppClient.RemoteError = RemoteError;
    AppClient.ApplicationError = ApplicationError;
    AppClient.SystemError = SystemError;
    return AppClient;
  },
  get AppServer() {
    return require("uu_appg01_core-appserver");
  },
  get Authentication() {
    let { Config } = require("uu_appg01_core-utils");
    let Authn = require("uu_appg01_core-authentication");
    if (!Config.hasKey(Authn.AuthenticationService.DEFAULT_PROVIDER_PARAM)) {
      // Configure default authentication service
      require("uu_appg01_oidc").OidcAuthenticationService;
      Config.set(Authn.AuthenticationService.DEFAULT_PROVIDER_PARAM, "uu_appg01_oidc");
    }
    return Authn;
  },
  get Authorization() {
    let { Config } = require("uu_appg01_core-utils");
    let Authz = require("uu_appg01_core-authorization");
    if (!Config.hasKey(Authz.AuthorizationService.DEFAULT_PROVIDER_PARAM)) {
      // Configure default authorization service
      require("uu_appg01_workspace").WorkspaceAuthorizationService;
      Config.set(Authz.AuthorizationService.DEFAULT_PROVIDER_PARAM, "uu_appg01_workspace");
    }
    return Authz;
  },
  get Logging() {
    return require("uu_appg01_core-logging");
  },
  get ObjectStore() {
    return require("uu_appg01_objectstore");
  },
  get Oidc() {
    return require("uu_appg01_oidc");
  },
  get Perflog() {
    const { Perflog, Section, SectionId, Logger } = require("uu_appg01_core-perflog");
    Perflog.Section = Section;
    Perflog.SectionId = SectionId;
    Perflog.Logger = Logger;
    return Perflog;
  },
  get Uri() {
    return require("uu_appg01_core-uri");
  },
  get Utils() {
    return require("uu_appg01_core-utils");
  },
  get Validation() {
    return require("uu_appg01_core-validation");
  },
  get Workspace() {
    return require("uu_appg01_workspace");
  },
  get AuditLog() {
    return require("uu_appg01_auditlog");
  },
  get Representation() {
    return require("uu_apprepresentationg01");
  }
};
