// uu_appg01_core-appserver/src/middleware/default-uc-handler.js
"use strict";

const { Config, OptsReader } = require("uu_appg01_core-utils");

const MIDDLEWARE_ORDER = -400;

/**
 * Enables redirection of base URL (URL which ends with tenant path part)
 * to specified default use case.<br>
 * Configuration parameters:<br>
 * <b>default_uc_path</b>
 * Default use case path relative to base URL.
 * <b>default_uc_use_redirect</b>
 * Flag whether use client redirection (true) or use internal forward.
 * Default to false.
 */
class DefaultUcHandler {
  constructor(opts = null) {
    this.order = MIDDLEWARE_ORDER;
    opts = new OptsReader(opts, Config);
    this.defaultUcPath = opts.getString("default_uc_path");
    this.defaultUcUseRedirect = opts.getBoolean("default_uc_use_redirect", true);
  }

  pre(req, res, next) {
    if (!this.defaultUcPath) {
      return next();
    }

    let path = req.url || "/";
    let pathParts = path.split("/");
    if (pathParts.length == 3 || (pathParts.length == 4 && pathParts[3] === "")) {
      pathParts[3] = this.defaultUcPath;
      let newPath = pathParts.join("/");

      if (this.defaultUcUseRedirect) {
        if (req.method.match(/^GET$/i)) {
          return res.redirect(302, newPath);
        } else {
          return res.redirect(309, newPath);
        }
      } else {
        res.locals._uu_app_default_uc_handler_orig_url = path;
        req.url = newPath;
      }
    }

    return next();
  }

  post(req, res, next) {
    if (res.locals._uu_app_default_uc_handler_orig_url) {
      req.url = res.locals._uu_app_default_uc_handler_orig_url;
    }
    return next();
  }

  onError(error, req, res, next) {
    if (res.locals._uu_app_default_uc_handler_orig_url) {
      req.url = res.locals._uu_app_default_uc_handler_orig_url;
    }
    return next(error);
  }
}

/**
 * Module exports Default UC handler constructor.
 * @type {DefaultUcHandler}
 */
module.exports = DefaultUcHandler;
