// uu5tilesg02/dist/mocks/animal-mock-data.mjs
export default [
  {
    id: "tygr01",
    speciesName: {
      cs: "Tygr indický",
      en: "Bengal tiger"
    },
    speciesTaxonomyName: "Panthera tigris",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Šelmy",
      en: "Carnivora"
    },
    family: {
      cs: "Kočkovití",
      en: "Felidae"
    },
    location: {
      cs: "Asie",
      en: "Asia"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "tygr_01"
  },
  {
    id: "pavian01",
    speciesName: {
      cs: "Pavián pláštíkový",
      en: "Hamadryas baboon"
    },
    speciesTaxonomyName: "Papio hamadryas",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Primáti",
      en: "Primates"
    },
    family: {
      cs: "Kočkodanovití",
      en: "Cercopithecidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "pavian_01"
  },
  {
    id: "malpa01",
    speciesName: {
      cs: "Malpa zlatobřichá",
      en: "Golden-bellied capuchin"
    },
    speciesTaxonomyName: "Sapajus xanthostemos",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Primáti",
      en: "Primates"
    },
    family: {
      cs: "Malpovití",
      en: "Cebidae"
    },
    location: {
      cs: "Amerika",
      en: "America"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "malpa_01"
  },
  {
    id: "langur01",
    speciesName: {
      cs: "Langur duk",
      en: "Red-shanked douc"
    },
    speciesTaxonomyName: "Pygathrix nemaeus",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Primáti",
      en: "Primates"
    },
    family: {
      cs: "Kočkodanovití",
      en: "Cercopithecidae"
    },
    location: {
      cs: "Asie",
      en: "Asia"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "langur_01"
  },
  {
    id: "orangutan01",
    speciesName: {
      cs: "Orangutan sumaterský",
      en: "Sumatran orangutan"
    },
    speciesTaxonomyName: "Pongo abelii",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Primáti",
      en: "Primates"
    },
    family: {
      cs: "Hominidi",
      en: "Hominidae"
    },
    location: {
      cs: "Asie",
      en: "Asia"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "orangutan_01"
  },
  {
    id: "varan02",
    speciesName: {
      cs: "Varan komodský",
      en: "Komodo dragon"
    },
    speciesTaxonomyName: "Varanus komodoensis",
    class1: {
      cs: "Plazi",
      en: "Reptilia"
    },
    order: {
      cs: "Šupinatí",
      en: "Squamata"
    },
    family: {
      cs: "Varanovití",
      en: "Varanidae"
    },
    location: {
      cs: "Asie",
      en: "Asia"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "varan_02"
  },
  {
    id: "krokodyl01",
    speciesName: {
      cs: "Krokodýl mořský",
      en: "Saltwater crocodile"
    },
    speciesTaxonomyName: "Crocodylus porosus",
    class1: {
      cs: "Plazi",
      en: "Reptilia"
    },
    order: {
      cs: "Krokodýli",
      en: "Crocodilia"
    },
    family: {
      cs: "Krokodýlovití",
      en: "Crocodylidae"
    },
    location: {
      cs: "Asie, Austrálie",
      en: "Asia, Australia"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "krokodyl_01"
  },
  {
    id: "mandril01",
    speciesName: {
      cs: "Mandril rýholící",
      en: "Mandrill"
    },
    speciesTaxonomyName: "Mandrillus sphinx",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Primáti",
      en: "Primates"
    },
    family: {
      cs: "Kočkodanovití",
      en: "Cercopithecidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "mandril_01"
  },
  {
    id: "simpanz01",
    speciesName: {
      cs: "Šimpanz učenlivý",
      en: "Common chimpanzee"
    },
    speciesTaxonomyName: "Pan troglodytes",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Primáti",
      en: "Primates"
    },
    family: {
      cs: "Hominidi",
      en: "Hominidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "simpanz_01"
  },
  {
    id: "kahau01",
    speciesName: {
      cs: "Kahau nosatý",
      en: "Proboscis monkey"
    },
    speciesTaxonomyName: "Nasalis larvatus",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Primáti",
      en: "Primates"
    },
    family: {
      cs: "Kočkodanovití",
      en: "Cercopithecidae"
    },
    location: {
      cs: "Asie",
      en: "Asia"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "kahau_01"
  },
  {
    id: "nosozec01",
    speciesName: {
      cs: "Nosorožec tuponosý",
      en: "White rhinoceros"
    },
    speciesTaxonomyName: "Ceratotherium simum",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Lichokopytníci",
      en: "Perissodactyla"
    },
    family: {
      cs: "Nosorožcovití",
      en: "Rhinocerotidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "nosorozec_01"
  },
  {
    id: "zelva02",
    speciesName: {
      cs: "Želva nádherná",
      en: "Red-eared slider"
    },
    speciesTaxonomyName: "Trachemys scripta",
    class1: {
      cs: "Plazi",
      en: "Reptilia"
    },
    order: {
      cs: "Želvy",
      en: "Testudines"
    },
    family: {
      cs: "Emidovití",
      en: "Emydidae"
    },
    location: {
      cs: "Amerika",
      en: "America"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "zelva_02"
  },
  {
    id: "skalnak01",
    speciesName: {
      cs: "Skalňák oranžový",
      en: "Guianan cock-of-the-rock"
    },
    speciesTaxonomyName: "Rupicola rupicola",
    class1: {
      cs: "Ptáci",
      en: "Aves"
    },
    order: {
      cs: "Pěvci",
      en: "Passeriformes"
    },
    family: {
      cs: "Kotingovití",
      en: "Cotingidae"
    },
    location: {
      cs: "Amerika",
      en: "America"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "skalnak_01"
  },
  {
    id: "gavial01",
    speciesName: {
      cs: "Gaviál indický",
      en: "Gharial"
    },
    speciesTaxonomyName: "Gavialus gangeticus",
    class1: {
      cs: "Plazi",
      en: "Reptilia"
    },
    order: {
      cs: "Krokodýli",
      en: "Crocodilia"
    },
    family: {
      cs: "Gaviálovití",
      en: "Gavialidae"
    },
    location: {
      cs: "Asie",
      en: "Asia"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "gavial_01"
  },
  {
    id: "zelva01",
    speciesName: {
      cs: "Batagur tuntong",
      en: "Southern river terrapin"
    },
    speciesTaxonomyName: "Batagur affinis",
    class1: {
      cs: "Plazi",
      en: "Reptilia"
    },
    order: {
      cs: "Želvy",
      en: "Testudines"
    },
    family: {
      cs: "Batagurovití",
      en: "Geoemydidae"
    },
    location: {
      cs: "Asie",
      en: "Asia"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "zelva_01"
  },
  {
    id: "panda02",
    speciesName: {
      cs: "Panda červená",
      en: "Red panda"
    },
    speciesTaxonomyName: "Ailurus fulgens",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Šelmy",
      en: "Carnivora"
    },
    family: {
      cs: "Pandy malé",
      en: "Ailuridae"
    },
    location: {
      cs: "Asie",
      en: "Asia"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "panda_02"
  },
  {
    id: "panda01",
    speciesName: {
      cs: "Panda velká",
      en: "Giant panda"
    },
    speciesTaxonomyName: "Ailuropoda melanoleuca",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Šelmy",
      en: "Carnivora"
    },
    family: {
      cs: "Medvědovití",
      en: "Ursidae"
    },
    location: {
      cs: "Asie",
      en: "Asia"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "panda_01"
  },
  {
    id: "kapybara01",
    speciesName: {
      cs: "Kapybara",
      en: "Capybara"
    },
    speciesTaxonomyName: "Hydrochoreus hydrochaeris",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Hlodavci",
      en: "Rodentia"
    },
    family: {
      cs: "Morčatovití",
      en: "Caviidae"
    },
    location: {
      cs: "Amerika",
      en: "America"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "kapybara_01"
  },
  {
    id: "tukan01",
    speciesName: {
      cs: "Tukan obrovský",
      en: "Toco toucan"
    },
    speciesTaxonomyName: "Ramphastos toco",
    class1: {
      cs: "Ptáci",
      en: "Aves"
    },
    order: {
      cs: "Šplhavci",
      en: "Piciformes"
    },
    family: {
      cs: "Tukanovití",
      en: "Ramphastidae"
    },
    location: {
      cs: "Amerika",
      en: "America"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "tukan_01"
  },
  {
    id: "aligator01",
    speciesName: {
      cs: "Aligátor čínský",
      en: "Chinese alligator"
    },
    speciesTaxonomyName: "Alligator sinensis",
    class1: {
      cs: "Plazi",
      en: "Reptilia"
    },
    order: {
      cs: "Krokodýli",
      en: "Crocodilia"
    },
    family: {
      cs: "Aligátorovití",
      en: "Alligatoridae"
    },
    location: {
      cs: "Asie",
      en: "Asia"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "aligator_01"
  },
  {
    id: "lachtan01",
    speciesName: {
      cs: "Lachtan jihoafrický",
      en: "Brown fur seal"
    },
    speciesTaxonomyName: "Arctocephalus pusillus",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Šelmy",
      en: "Carnivora"
    },
    family: {
      cs: "Lachtanovití",
      en: "Otariidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "lachtan_01"
  },
  {
    id: "plamenak01",
    speciesName: {
      cs: "Plameňák růžový",
      en: "Greater flamingo"
    },
    speciesTaxonomyName: "Phoenicopterus roseus",
    class1: {
      cs: "Ptáci",
      en: "Aves"
    },
    order: {
      cs: "Plameňáci",
      en: "Phoenicopteriformes"
    },
    family: {
      cs: "Plameňákovití",
      en: "Phoenicopteridae"
    },
    location: {
      cs: "Evropa, Asie, Afrika",
      en: "Europe, Asia, Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "plamenak_01"
  },
  {
    id: "chameleon01",
    speciesName: {
      cs: "Chameleon namibijský",
      en: "Namaqua chameleon"
    },
    speciesTaxonomyName: "Chameleo namaquensis",
    class1: {
      cs: "Plazi",
      en: "Reptilia"
    },
    order: {
      cs: "Šupinatí",
      en: "Squamata"
    },
    family: {
      cs: "Chameleonovití",
      en: "Chamaeleonidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "chameleon_01"
  },
  {
    id: "agama01",
    speciesName: {
      cs: "Agama osadní",
      en: "Agama agama"
    },
    speciesTaxonomyName: "Agama agama",
    class1: {
      cs: "Plazi",
      en: "Reptilia"
    },
    order: {
      cs: "Šupinatí",
      en: "Squamata"
    },
    family: {
      cs: "Agamovití",
      en: "Agamidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "agama_01"
  },
  {
    id: "hyena01",
    speciesName: {
      cs: "Hyena skvrnitá",
      en: "Spotted hyena"
    },
    speciesTaxonomyName: "Crocuta crocuta",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Šelmy",
      en: "Carnivora"
    },
    family: {
      cs: "Hyenovití",
      en: "Hyaenidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "hyena_01"
  },
  {
    id: "sakal01",
    speciesName: {
      cs: "Šakal čabrakový",
      en: "Black-backed jackal"
    },
    speciesTaxonomyName: "Canis mesomelas",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Šelmy",
      en: "Carnivora"
    },
    family: {
      cs: "Psovití",
      en: "Canidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "sakal_01"
  },
  {
    id: "pstros01",
    speciesName: {
      cs: "Pštros dvouprstý",
      en: "Common ostrich"
    },
    speciesTaxonomyName: "Struthio camelus",
    class1: {
      cs: "Ptáci",
      en: "Aves"
    },
    order: {
      cs: "Pštrosi",
      en: "Struthioniformes"
    },
    family: {
      cs: "Pštrosovití",
      en: "Struthionidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "pstros_01"
  },
  {
    id: "hadilov01",
    speciesName: {
      cs: "Hadilov písař",
      en: "Secretarybird"
    },
    speciesTaxonomyName: "Sagittarius serpentarius",
    class1: {
      cs: "Ptáci",
      en: "Aves"
    },
    order: {
      cs: "Dravci",
      en: "Accipitriformes"
    },
    family: {
      cs: "Hadilovovití",
      en: "Sagittariidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "hadilov_01"
  },
  {
    id: "kudu01",
    speciesName: {
      cs: "Kudu velký",
      en: "Greater kudu"
    },
    speciesTaxonomyName: "Tragelaphus strepsiceros",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Sudokopytníci",
      en: "Artiodactyla"
    },
    family: {
      cs: "Turovití",
      en: "Bovidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "kudu_01"
  },
  {
    id: "mamba01",
    speciesName: {
      cs: "Mamba černá",
      en: "Black mamba"
    },
    speciesTaxonomyName: "Dendroaspis polylepis",
    class1: {
      cs: "Plazi",
      en: "Reptilia"
    },
    order: {
      cs: "Šupinatí",
      en: "Squamata"
    },
    family: {
      cs: "Korálovcovití",
      en: "Elapidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "mamba_01"
  },
  {
    id: "zebra01",
    speciesName: {
      cs: "Zebra Burchellova",
      en: "Burchell's zebra"
    },
    speciesTaxonomyName: "Equus quagga burchellii",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Lichokopytníci",
      en: "Perissodactyla"
    },
    family: {
      cs: "Koňovití",
      en: "Equidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "zebra_01"
  },
  {
    id: "dikdik01",
    speciesName: {
      cs: "Dikdik Kirkův",
      en: "Kirk's dik-dik"
    },
    speciesTaxonomyName: "Madoqua kirkii",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Sudokopytníci",
      en: "Artiodactyla"
    },
    family: {
      cs: "Turovití",
      en: "Bovidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "dikdik_01"
  },
  {
    id: "pakun01",
    speciesName: {
      cs: "Pakůň žíhaný",
      en: "Blue wildebeest"
    },
    speciesTaxonomyName: "Connochaetes taurinus",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Sudokopytníci",
      en: "Artiodactyla"
    },
    family: {
      cs: "turovití",
      en: "Bovidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "pakun_01"
  },
  {
    id: "buvolec01",
    speciesName: {
      cs: "Buvolec modrý",
      en: "Common tsessebe"
    },
    speciesTaxonomyName: "Damaliscus lunatus",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Sudokopytníci",
      en: "Artiodactyla"
    },
    family: {
      cs: "Turovití",
      en: "Bovidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "buvolec_01"
  },
  {
    id: "impala01",
    speciesName: {
      cs: "Impala",
      en: "Impala"
    },
    speciesTaxonomyName: "Aepyceros melampus",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Sudokopytníci",
      en: "Artiodactyla"
    },
    family: {
      cs: "Turovití",
      en: "Bovidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "impala_01"
  },
  {
    id: "vazka02",
    speciesName: {
      cs: "Vážka",
      en: "Yellow-winged darter"
    },
    speciesTaxonomyName: "Trithemis kirbyi",
    class1: {
      cs: "Hmyz",
      en: "Insecta"
    },
    order: {
      cs: "Vážky",
      en: "Odonata"
    },
    family: {
      cs: "Vážkovití",
      en: "Libellulidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "vazka_02"
  },
  {
    id: "voduska01",
    speciesName: {
      cs: "Voduška červená",
      en: "Lechwe"
    },
    speciesTaxonomyName: "Kobus leche",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Sudokopytníci",
      en: "Artiodactyla"
    },
    family: {
      cs: "Turovití",
      en: "Bovidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "voduska_01"
  },
  {
    id: "zoborozec01",
    speciesName: {
      cs: "Zoborožec jihoafrický",
      en: "Southern yellow-billed hornbill"
    },
    speciesTaxonomyName: "Tckus leucomelas",
    class1: {
      cs: "Ptáci",
      en: "Aves"
    },
    order: {
      cs: "Zoborožci",
      en: "Bucerotiformes"
    },
    family: {
      cs: "Zoborožcovití",
      en: "Bucerotidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "zoborozec_01"
  },
  {
    id: "vlha01",
    speciesName: {
      cs: "Vlha běločelá",
      en: "White-fronted bee-eater"
    },
    speciesTaxonomyName: "Merops bullockoides",
    class1: {
      cs: "Ptáci",
      en: "Aves"
    },
    order: {
      cs: "Srostloprstí",
      en: "Coraciiformes"
    },
    family: {
      cs: "Vlhovití",
      en: "Meropidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "vlha_01"
  },
  {
    id: "lednacek01",
    speciesName: {
      cs: "Rybařík jižní",
      en: "Pied kingfisher"
    },
    speciesTaxonomyName: "Ceryle rudis",
    class1: {
      cs: "Ptáci",
      en: "Aves"
    },
    order: {
      cs: "Srostloprstí",
      en: "Coraciiformes"
    },
    family: {
      cs: "Ledňáčkovití",
      en: "Alcedinidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "lednacek_01"
  },
  {
    id: "varan01",
    speciesName: {
      cs: "Varan nilský",
      en: "Nile monitor"
    },
    speciesTaxonomyName: "Varanus niloticus",
    class1: {
      cs: "Plazi",
      en: "Reptilia"
    },
    order: {
      cs: "Šupinatí",
      en: "Squamata"
    },
    family: {
      cs: "Varanovití",
      en: "Varanidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "varan_01"
  },
  {
    id: "slon01",
    speciesName: {
      cs: "Slon africký",
      en: "African bush elephant"
    },
    speciesTaxonomyName: "Loxodonta africana",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Chobotnatci",
      en: "Proboscidea"
    },
    family: {
      cs: "Slonovití",
      en: "Elephantidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "slon_01"
  },
  {
    id: "nesyt01",
    speciesName: {
      cs: "Nesyt africký",
      en: "Yellow-billed stork"
    },
    speciesTaxonomyName: "Mycteria ibis",
    class1: {
      cs: "Ptáci",
      en: "Aves"
    },
    order: {
      cs: "Brodiví",
      en: "Ciconiiformes"
    },
    family: {
      cs: "Čápovití",
      en: "Ciconiidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "nesyt_01"
  },
  {
    id: "perlicka01",
    speciesName: {
      cs: "Perlička kropenatá",
      en: "Helmeted guineafowl"
    },
    speciesTaxonomyName: "Numida meleagris",
    class1: {
      cs: "Ptáci",
      en: "Aves"
    },
    order: {
      cs: "Hrabaví",
      en: "Galliformes"
    },
    family: {
      cs: "Perličkovití",
      en: "Numididae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "perlicka_01"
  },
  {
    id: "buvol01",
    speciesName: {
      cs: "Buvol africký",
      en: "African buffalo"
    },
    speciesTaxonomyName: "Syncerus cafer",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Sudokopytníci",
      en: "Artiodactyla"
    },
    family: {
      cs: "Turovití",
      en: "Bovidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "buvol_01"
  },
  {
    id: "mandelik01",
    speciesName: {
      cs: "Mandelík fialovoprsý",
      en: "Lilac-breasted roller"
    },
    speciesTaxonomyName: "Coracias caudatus",
    class1: {
      cs: "Ptáci",
      en: "Aves"
    },
    order: {
      cs: "Srostloprstí",
      en: "Coraciiformes"
    },
    family: {
      cs: "Mandelíkovití",
      en: "Coraciidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "mandelik_01"
  },
  {
    id: "hroch01",
    speciesName: {
      cs: "Hroch obojživelný",
      en: "Hippopotamus"
    },
    speciesTaxonomyName: "Hippopotamus amphibius",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Sudokopytníci",
      en: "Artiodactyla"
    },
    family: {
      cs: "Hrochovití",
      en: "Hippopotamidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "hroch_01"
  },
  {
    id: "kockodan01",
    speciesName: {
      cs: "Kočkodan zelený",
      en: "Green monkey"
    },
    speciesTaxonomyName: "Chlorocebus sabaeus",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Primáti",
      en: "Primates"
    },
    family: {
      cs: "Kočkodanovití",
      en: "Cercopithecidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "kockodan_01"
  },
  {
    id: "prase01",
    speciesName: {
      cs: "Prase bradavičnaté",
      en: "Desert warthog"
    },
    speciesTaxonomyName: "Phacochoerus aethiopicus",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Sudokopytníci",
      en: "Artiodactyla"
    },
    family: {
      cs: "Prasatovití",
      en: "Suidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "prase_01"
  },
  {
    id: "slon02",
    speciesName: {
      cs: "Slon africký",
      en: "African bush elephant"
    },
    speciesTaxonomyName: "Loxodonta africana",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Chobotnatci",
      en: "Proboscidea"
    },
    family: {
      cs: "Slonovití",
      en: "Elephantidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "slon_02"
  },
  {
    id: "sup01",
    speciesName: {
      cs: "Sup kapský",
      en: "Cape vulture"
    },
    speciesTaxonomyName: "Gyps coprotheres",
    class1: {
      cs: "Ptáci",
      en: "Aves"
    },
    order: {
      cs: "Dravci",
      en: "Accipitriformes"
    },
    family: {
      cs: "Jestřábovití",
      en: "Accipitridae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "sup_01"
  },
  {
    id: "lev01",
    speciesName: {
      cs: "Lev pustinný",
      en: ""
    },
    speciesTaxonomyName: "Panthera leo",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Šelmy",
      en: "Carnivora"
    },
    family: {
      cs: "Kočkovití",
      en: "Felidae"
    },
    location: {
      cs: "Afrika, Asie",
      en: "Africa, Asia"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "lev_01"
  },
  {
    id: "krab01",
    speciesName: {
      cs: "Krab",
      en: "Round-eyed ghost crab"
    },
    speciesTaxonomyName: "Ocypode cordimana",
    class1: {
      cs: "Rakovci",
      en: "Malacostraca"
    },
    order: {
      cs: "Desetinožci",
      en: "Decapoda"
    },
    family: {
      cs: "Ocypodidae",
      en: "Ocypodidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "krab_01"
  },
  {
    id: "kalon01",
    speciesName: {
      cs: "Kaloň seychelský",
      en: "Seychelles fruit bat"
    },
    speciesTaxonomyName: "Pteropus seychellensis",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Letouni",
      en: "Chiroptera"
    },
    family: {
      cs: "Kaloňovití",
      en: "Pteropodidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "kalon_01"
  },
  {
    id: "ryba02",
    speciesName: {
      cs: "Pomec azurový",
      en: "Koran angelfish"
    },
    speciesTaxonomyName: "Pomacanthus semicirculatus",
    class1: {
      cs: "Paprskoploutví",
      en: "Actinopterygii"
    },
    order: {
      cs: "Ostnoploutví",
      en: "Perciformes"
    },
    family: {
      cs: "Pomcovití",
      en: "Pomacanthidae"
    },
    location: {
      cs: "Indický oceán",
      en: "Indian Ocean"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "ryba_02"
  },
  {
    id: "ryba01",
    speciesName: {
      cs: "Kněžík paví",
      en: "Ornate wrasse"
    },
    speciesTaxonomyName: "Thalassoma pavo",
    class1: {
      cs: "Paprskoploutví",
      en: "Actinopterygii"
    },
    order: {
      cs: "Ostnoploutví",
      en: "Perciformes"
    },
    family: {
      cs: "Pyskounovití",
      en: "Labridae"
    },
    location: {
      cs: "Indický oceán",
      en: "Indian Ocean"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "ryba_01"
  },
  {
    id: "rejnok01",
    speciesName: {
      cs: "Trnucha indická",
      en: "Porcupine ray"
    },
    speciesTaxonomyName: "Urogymnus asperrimus",
    class1: {
      cs: "Paryby",
      en: "Chondrichthyes"
    },
    order: {
      cs: "Rejnoci",
      en: "Myliobatiformes"
    },
    family: {
      cs: "Trnuchovití",
      en: "Dasyatidae"
    },
    location: {
      cs: "Indický oceán",
      en: "Indian Ocean"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "rejnok_01"
  },
  {
    id: "strdimil01",
    speciesName: {
      cs: "Strdimil seychelský",
      en: "Seychelles sunbird"
    },
    speciesTaxonomyName: "Cinnyris dussumieri",
    class1: {
      cs: "Ptáci",
      en: "Aves"
    },
    order: {
      cs: "Pěvci",
      en: "Passeriformes"
    },
    family: {
      cs: "Strdimilovití",
      en: "Nectariniidae"
    },
    location: {
      cs: "Afrika",
      en: "Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "strdimil_01"
  },
  {
    id: "kareta01",
    speciesName: {
      cs: "Kareta pravá",
      en: "Hawksbill sea turtle"
    },
    speciesTaxonomyName: "Eretmochelys imbricata",
    class1: {
      cs: "Plazi",
      en: "Reptilia"
    },
    order: {
      cs: "Želvy",
      en: "Testudines"
    },
    family: {
      cs: "Karetovití",
      en: "Cheloniidae"
    },
    location: {
      cs: "Oceány",
      en: "Oceans"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "kareta_01"
  },
  {
    id: "krab02",
    speciesName: {
      cs: "Krab",
      en: "Cardisoma carnifex"
    },
    speciesTaxonomyName: "Cardisoma carnifex",
    class1: {
      cs: "Rakovci",
      en: "Malacostraca"
    },
    order: {
      cs: "Desetinožci",
      en: "Decapoda"
    },
    family: {
      cs: "Gecarcinidae",
      en: "Gecarcinidae"
    },
    location: {
      cs: "Afrika, Asie",
      en: "Africa, Asia"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "krab_02"
  },
  {
    id: "vazka01",
    speciesName: {
      cs: "Vážka čtyřskvrnná",
      en: "Four-spotted chaser"
    },
    speciesTaxonomyName: "Libellula quadrimaculata",
    class1: {
      cs: "Hmyz",
      en: "Insecta"
    },
    order: {
      cs: "Vážky",
      en: "Odonata"
    },
    family: {
      cs: "Vážkovití",
      en: "Libellulidae"
    },
    location: {
      cs: "Evropa",
      en: "Europa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "vazka_01"
  },
  {
    id: "skokan01",
    speciesName: {
      cs: "Skokan skřehotavý",
      en: "Marsh frog"
    },
    speciesTaxonomyName: "Pelophylax ridibundus",
    class1: {
      cs: "Obojživelníci",
      en: "Amphibia"
    },
    order: {
      cs: "Žáby",
      en: "Anura"
    },
    family: {
      cs: "Skokanovití",
      en: "Ranidae"
    },
    location: {
      cs: "Evropa, Asie",
      en: "Europe, Asia"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "skokan_01"
  },
  {
    id: "kolpik01",
    speciesName: {
      cs: "Kolpík bílý",
      en: "Eurasian spoonbill"
    },
    speciesTaxonomyName: "Platalea leucorodia",
    class1: {
      cs: "Ptáci",
      en: "Aves"
    },
    order: {
      cs: "Brodiví",
      en: "Pelecaniformes"
    },
    family: {
      cs: "Ibisovití",
      en: "Threskiornithidae"
    },
    location: {
      cs: "Evropa, Asie, Afrika",
      en: "Europe, Asia, Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "kolpik_01"
  },
  {
    id: "nosal01",
    speciesName: {
      cs: "Nosál bělohubý",
      en: "White-nosed coati"
    },
    speciesTaxonomyName: "Nasua narica",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Šelmy",
      en: "Carnivora"
    },
    family: {
      cs: "Medvídkovití",
      en: "Procyonidae"
    },
    location: {
      cs: "Amerika",
      en: "America"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "nosal_01"
  },
  {
    id: "tapir01",
    speciesName: {
      cs: "Tapír čabrakový",
      en: "Malayan tapir"
    },
    speciesTaxonomyName: "Tapirus indicus",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Lichokopytníci",
      en: "Perissodactyla"
    },
    family: {
      cs: "Tapírovití",
      en: "Tapiridae"
    },
    location: {
      cs: "Asie",
      en: "Asia"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "tapir_01"
  },
  {
    id: "volavka01",
    speciesName: {
      cs: "Volavka popelavá",
      en: "Grey heron"
    },
    speciesTaxonomyName: "Ardea cinerea",
    class1: {
      cs: "Ptáci",
      en: "Aves"
    },
    order: {
      cs: "Veslonozí",
      en: "Pelecaniformes"
    },
    family: {
      cs: "Volavkovití",
      en: "Ardeidae"
    },
    location: {
      cs: "Evropa, Asie, Afrika",
      en: "Europe, Asia, Africa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "volavka_01"
  },
  {
    id: "motyl02",
    speciesName: {
      cs: "Otakárek",
      en: "Ambrax butterfly"
    },
    speciesTaxonomyName: "Papilio ambrax",
    class1: {
      cs: "Hmyz",
      en: "Insecta"
    },
    order: {
      cs: "Motýli",
      en: "Lepidoptera"
    },
    family: {
      cs: "Otakárkovití",
      en: "Papilionidae"
    },
    location: {
      cs: "Austrálie",
      en: "Australia"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "motyl_02"
  },
  {
    id: "motyl01",
    speciesName: {
      cs: "Motýl",
      en: "Paper kite"
    },
    speciesTaxonomyName: "Idea leuconoe",
    class1: {
      cs: "Hmyz",
      en: "Insecta"
    },
    order: {
      cs: "Motýli",
      en: "Lepidoptera"
    },
    family: {
      cs: "Babočkovití",
      en: "Nymphalidae"
    },
    location: {
      cs: "Asie, Austrálie",
      en: "Asia, Australia"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "motyl_01"
  },
  {
    id: "lvicek01",
    speciesName: {
      cs: "Lvíček zlatohlavý",
      en: "Golden-headed lion tamarin"
    },
    speciesTaxonomyName: "Leontopithecus chrysomelas",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Primáti",
      en: "Primates"
    },
    family: {
      cs: "Malpovití",
      en: "Callitrichidae"
    },
    location: {
      cs: "Amerika",
      en: "America"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "lvicek_01"
  },
  {
    id: "pralesnicka01",
    speciesName: {
      cs: "Pralesnička harlekýn",
      en: "Yellow-banded poison dart frog"
    },
    speciesTaxonomyName: "Dendrobates leucomelas",
    class1: {
      cs: "Obojživelníci",
      en: "Amphibia"
    },
    order: {
      cs: "Žáby",
      en: "Anura"
    },
    family: {
      cs: "Pralesničkovití",
      en: "Dendrobatidae"
    },
    location: {
      cs: "Amerika",
      en: "America"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "pralesnicka_01"
  },
  {
    id: "pirana01",
    speciesName: {
      cs: "Piraňa obecná",
      en: "Red-bellied piranha"
    },
    speciesTaxonomyName: "Pygocentrus nattereri",
    class1: {
      cs: "Paprskoploutví",
      en: "Actinopterygii"
    },
    order: {
      cs: "Trnobřiší",
      en: "Characiformes"
    },
    family: {
      cs: "Tetrovití",
      en: "Serrasalmidae"
    },
    location: {
      cs: "Amerika",
      en: "America"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "pirana_01"
  },
  {
    id: "klaun01",
    speciesName: {
      cs: "Klaun špičatopruhý",
      en: "Red Sea clownfish"
    },
    speciesTaxonomyName: "Amphiprion bicinctus",
    class1: {
      cs: "Paprskoploutví",
      en: "Actinopterygii"
    },
    order: {
      cs: "Ostnoploutví",
      en: "Perciformes"
    },
    family: {
      cs: "Sapínovití",
      en: "Pomacentridae"
    },
    location: {
      cs: "Indický oceán",
      en: "Indian Ocean"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "klaun_01"
  },
  {
    id: "kozoroh01",
    speciesName: {
      cs: "Kozorožec horský",
      en: "Alpine ibex"
    },
    speciesTaxonomyName: "Capra ibex",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Sudokopytníci",
      en: "Artiodactyla"
    },
    family: {
      cs: "Turovití",
      en: "Bovidae"
    },
    location: {
      cs: "Evropa",
      en: "Europa"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "kozoroh_01"
  },
  {
    id: "irbis01",
    speciesName: {
      cs: "Irbis",
      en: "Snow leopard"
    },
    speciesTaxonomyName: "Panthera uncia",
    class1: {
      cs: "Savci",
      en: "Mammalia"
    },
    order: {
      cs: "Šelmy",
      en: "Carnivora"
    },
    family: {
      cs: "Kočkovití",
      en: "Felidae"
    },
    location: {
      cs: "Asie",
      en: "Asia"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "irbis_01"
  },
  {
    id: "sovice01",
    speciesName: {
      cs: "Sovice sněžní",
      en: "Snowy owl"
    },
    speciesTaxonomyName: "Bubo scandiacus",
    class1: {
      cs: "Ptáci",
      en: "Aves"
    },
    order: {
      cs: "Sovy",
      en: "Strigiformes"
    },
    family: {
      cs: "Puštíkovití",
      en: "Strigidae"
    },
    location: {
      cs: "Evropa, Asie, Amerika",
      en: "Europe, Asia, America"
    },
    description: {
      cs: "",
      en: ""
    },
    img: "sovice_01"
  }
];
