// uu_appg01_workspace/src/api/controllers/sys-app-workspace-controller.js
"use strict";

const SysAppWorkspaceAbl = require("../../abl/sys-app-workspace-abl.js");
const AppClientTokenService = require("../../app-client-token-service");

class SysAppWorkspaceController {
  create(ucEnv) {
    return SysAppWorkspaceAbl.create({
      ucUri: ucEnv.uri,
      dtoIn: ucEnv.getDtoIn(),
      authzResult: ucEnv.getAuthorizationResult(),
      uuIdentity: ucEnv
        .getSession()
        .getIdentity()
        .getUuIdentity(),
      appClientToken: AppClientTokenService.getToken(ucEnv)
    });
  }

  get(ucEnv) {
    const awid = ucEnv.uri.getAwid();
    return SysAppWorkspaceAbl.get(awid);
  }

  set(ucEnv) {
    const awid = ucEnv.uri.getAwid();
    const dtoIn = ucEnv.getDtoIn();
    const session = ucEnv.getSession();
    return SysAppWorkspaceAbl.set(awid, dtoIn, session.getIdentity().getUuIdentity());
  }

  close(ucEnv) {
    const dtoIn = ucEnv.getDtoIn();
    return SysAppWorkspaceAbl.close(dtoIn);
  }

  suspend(ucEnv) {
    const dtoIn = ucEnv.getDtoIn();
    return SysAppWorkspaceAbl.suspend(dtoIn);
  }

  resume(ucEnv) {
    const dtoIn = ucEnv.getDtoIn();
    return SysAppWorkspaceAbl.resume(dtoIn);
  }

  setActiveSysState(ucEnv) {
    const awid = ucEnv.uri.getAwid();
    const dtoIn = ucEnv.getDtoIn();
    const session = ucEnv.getSession();
    return SysAppWorkspaceAbl.setActiveSysState(awid, dtoIn, session.getIdentity().getUuIdentity());
  }

  setRestrictedSysState(ucEnv) {
    const awid = ucEnv.uri.getAwid();
    const dtoIn = ucEnv.getDtoIn();
    const session = ucEnv.getSession();
    return SysAppWorkspaceAbl.setRestrictedSysState(awid, dtoIn, session.getIdentity().getUuIdentity());
  }

  setReadOnlySysState(ucEnv) {
    const awid = ucEnv.uri.getAwid();
    const dtoIn = ucEnv.getDtoIn();
    const session = ucEnv.getSession();
    return SysAppWorkspaceAbl.setReadOnlySysState(awid, dtoIn, session.getIdentity().getUuIdentity());
  }

  getLicenseOwner(ucEnv) {
    const awid = ucEnv.uri.getAwid();
    return SysAppWorkspaceAbl.getLicenseOwner(awid);
  }

  setAuthorizationStrategy(ucEnv) {
    const uri = ucEnv.uri;
    const dtoIn = ucEnv.getDtoIn();
    const session = ucEnv.getSession();
    return SysAppWorkspaceAbl.setAuthorizationStrategy(uri, dtoIn, session);
  }

  connectArtifact(ucEnv) {
    const uri = ucEnv.uri;
    const dtoIn = ucEnv.getDtoIn();
    const session = ucEnv.getSession();
    return SysAppWorkspaceAbl.connectArtifact(uri, dtoIn, session);
  }

  delete(ucEnv) {
    const dtoIn = ucEnv.getDtoIn();
    return SysAppWorkspaceAbl.delete(dtoIn);
  }

  list(ucEnv) {
    const dtoIn = ucEnv.getDtoIn();
    return SysAppWorkspaceAbl.list(dtoIn);
  }
}

module.exports = new SysAppWorkspaceController();
