// uu_appg01_workspace/src/middleware/app-workspace-init-middleware.js
"use strict";

const { Config } = require("uu_appg01_core-utils");
const { LoggerFactory } = require("uu_appg01_core-logging");
const { DaoFactory } = require("uu_appg01_datastore");

const AppWorkspaceComponent = require("../api/components/app-workspace");

const MIDDLEWARE_ORDER = 250;

class AppWorkspaceInitMiddleware {
  constructor() {
    this.logger = LoggerFactory.get("AppWorkspaceInitMiddleware");
    this.name = "App Workspace init Handler";
    // noinspection JSUnusedGlobalSymbols
    this.order = MIDDLEWARE_ORDER;
    this.asid = Config.get("asid");
    this.isDataStoreOn = DaoFactory.isDataStoreOn();
  }

  async post(req, res, next) {
    if (!this.isDataStoreOn) {
      return next();
    }

    const uri = res.locals.uri;
    if (uri.getAwid() !== this.asid && uri.getUseCase() === "sys/uuAppWorkspace/init") {
      let workspace = await AppWorkspaceComponent.get(uri.getAwid());
      if (workspace.sysState === "created") {
        await AppWorkspaceComponent.setActiveSysState(uri.getAwid());
        //TODO baseregistry
      }
    }
    next();
  }
}

module.exports = AppWorkspaceInitMiddleware;
