// uu_appg01_devkit-common/src/templates/__uu5-app/src/config/lsi.js
const Lsi = {
  appName: {
    cs: "Aplikace <%= context.vendor + context.app.replace(/^./, m => m.toUpperCase()).replace(/g[0-9]+$/, '') %>",
    en: "Application <%= context.vendor + context.app.replace(/^./, m => m.toUpperCase()).replace(/g[0-9]+$/, '') %>",
  },

  left: {
    home: {
      cs: "Vítejte",
      en: "Welcome",
    },
    about: {
      cs: "O aplikaci",
      en: "About Application",
    },
  },

  about: {
    header: {
      cs: "O aplikaci <%= context.vendor + context.app.replace(/^./, m => m.toUpperCase()).replace(/g[0-9]+$/, '') %>",
      en: "About application <%= context.vendor + context.app.replace(/^./, m => m.toUpperCase()).replace(/g[0-9]+$/, '') %>",
    },
    creatorsHeader: {
      cs: "Tvůrci aplikace",
      en: "Application creators",
    },
    termsOfUse: {
      cs: "Podmínky užívání",
      en: "Terms of use",
    },
  },

  bottom: {
    termsOfUse: {
      cs: "Podmínky užívání",
      en: "Terms of use",
    },
  },

  auth: {
    welcome: {
      cs: "Vítejte",
      en: "Welcome",
    },

    intro: {
      cs: `<uu5string/>Tato šablona obsahuje připravenou klientskou a serverovou část. Jednotlivé komponety, které jsou zde zobrazeny,
          jsou určeny k tomu, aby demonstrovaly možnosti a způsob použití. Je vhodné je upravit, zkopírovat či smazat pro
          potřeby vyvíjené aplikace. Více o struktuře uuApp se dozvíte v dokumetaci viz&nbsp;
          <UU5.Bricks.Link
            href="https://uuapp.plus4u.net/uu-bookkit-maing01/e884539c8511447a977c7ff070e7f2cf/book"
            target="_blank"
            content="uuAppDevKit"
          />.`,
      en: `<uu5string/>This template consist of prepared client and server side. Shown components demonstrate possibilities and way of
          using. For application developing purposes they are suitable for modifying, copying and deleting. More about
          uuApp Structure see documentation&nbsp;
          <UU5.Bricks.Link
            href="https://uuapp.plus4u.net/uu-bookkit-maing01/e884539c8511447a977c7ff070e7f2cf/book/page?code=stepByStepApp"
            target="_blank"
            content="uuAppDevKit"
          />.`,
    },

    clientSide: {
      cs: `<uu5string/>Klientská část je implementovaná s využitím komponent z knihoven <UU5.Bricks.LinkUU5 /> a <UU5.Bricks.LinkUuPlus4U5 />.`,
      en: `<uu5string/>Libraries <UU5.Bricks.LinkUU5/> and <UU5.Bricks.LinkUuPlus4U5 /> are used for developing of client side.`,
    },

    serverSide: {
      cs: `<uu5string/>Pro spuštení serverové části je potřeba provést inicializaci workspace podle návodu viz
          <UU5.Bricks.Link
            href="https://uuapp.plus4u.net/uu-bookkit-maing01/e884539c8511447a977c7ff070e7f2cf/book/page?code=stepByStepApp"
            target="_blank"
            content="uuApp Template Developer Guide"
          />.`,
      en: `<uu5string/>It is necessary to initialize application workspace for running server side. See manual
          <UU5.Bricks.Link
            href="https://uuapp.plus4u.net/uu-bookkit-maing01/e884539c8511447a977c7ff070e7f2cf/book/page?code=stepByStepApp"
            target="_blank"
            content="uuApp Template Developer Guide"
          />.`,
    },
  },

  unauth: {
    continueToMain: {
      cs: "Pokračovat na web produktu",
      en: "Continue to the product web",
    },
    notAuthorized: {
      cs: "Nemáte dostatečná práva k použití aplikace",
      en: "You do not have sufficient rights to use the application",
    },
  },

  unauthInit: {
    buyYourOwn: {
      cs: "Můžete si koupit vlastní <%= context.vendor + context.app.replace(/^./, m => m.toUpperCase()).replace(/g[0-9]+$/, '') %>.",
      en: "You can buy your own <%= context.vendor + context.app.replace(/^./, m => m.toUpperCase()).replace(/g[0-9]+$/, '') %>.",
    },
    notAuthorized: {
      cs: "Nemáte právo inicializovat tuto aplikaci <%= context.vendor + context.app.replace(/^./, m => m.toUpperCase()).replace(/g[0-9]+$/, '') %>.",
      en: "You don't have rights to initialize this <%= context.vendor + context.app.replace(/^./, m => m.toUpperCase()).replace(/g[0-9]+$/, '') %>.",
    },
  },

  controlPanel: {
    rightsError: {
      cs: "K zobrazení komponenty nemáte dostatečná práva.",
      en: "You do not have sufficient rights to display this component.",
    },

    btNotConnected: {
      cs: "Aplikace není napojená na Business Territory",
      en: "The application is not connected to a Business Territory",
    },
  },
};

export default Lsi;
