// uu_appg01_oidc/src/browser/ui/dom.js
let isBrowser = new Function("try {return this===window;}catch(e){ return false;}")();
let domReadyPromise = isBrowser ? (document.readyState === "interactive" || document.readyState === "complete" ? Promise.resolve() : new Promise((resolve) => {
  document.addEventListener("DOMContentLoaded", resolve, false);
})) : Promise.resolve();

const WINDOW = location.href.startsWith("about:") && window.frameElement ? parent : window;
const BASE_ELEMENT = WINDOW.document.querySelector("base");
const APP_BASE_URI = WINDOW.location.protocol + "//" + WINDOW.location.host + ((BASE_ELEMENT && BASE_ELEMENT.getAttribute("data-uu-app-base") || "") || "/");
const CANONICAL_APP_BASE_URI = ((new RegExp("uu.app.cbu=([^;]+)")).exec(WINDOW.document.cookie) || [])[1];

module.exports = {
  domReady: domReadyPromise,
  window: WINDOW,
  baseElement: BASE_ELEMENT,
  appBaseUri: APP_BASE_URI,
  canonicalAppBaseUri: CANONICAL_APP_BASE_URI,
  toFullAppUrl: function(path) {
    // relative & absolute paths are relative to app base
    let a = WINDOW.document.createElement("a");
    a.href = path && path.match(/^[a-z]+:/i) ? path : APP_BASE_URI + (path ? path.replace(/^\/+/, "") : "");
    return a.href.toString(); // browser-normalized URL (removed "../" sequences, ...)
  }
};
