// uu_appg01_objectstore/index.js
"use strict";

module.exports = {
  get DaoFactory() {
    return require("uu_appg01_datastore").DaoFactory;
  },
  get Lock() {
    return require("./src/libs/objectstore-lock");
  },
  get UuObjectDao() {
    return require("./src/libs/dao");
  },
  get ObjectStoreError() {
    return require("./src/dao_errors/objectstore-error");
  },
  get DuplicateKey() {
    return require("./src/dao_errors/duplicate-key.js");
  },
  get ExpiredLock() {
    return require("./src/dao_errors/expired-lock.js");
  },
  get InvalidRevision() {
    return require("./src/dao_errors/invalid-revision.js");
  },
  get LockDoesNotExist() {
    return require("./src/dao_errors/lock-does-not-exist.js");
  },
  get MissingRevision() {
    return require("./src/dao_errors/missing-revision.js");
  },
  get NoiExceeced() {
    return require("./src/dao_errors/noi-exceeded.js");
  },
  get ObjectLocked() {
    return require("./src/dao_errors/object-locked.js");
  },
  get ObjectNotFound() {
    return require("./src/dao_errors/object-not-found.js");
  },
  get SoiExceeded() {
    return require("./src/dao_errors/soi-exceeded.js");
  },
  get UnexpectedError() {
    return require("./src/dao_errors/unexpected-error.js");
  },
};
