// uu_appg01_devkit-common/src/upgrades/upgrade-4.3.0.js
const fs = require("fs-extra");
const path = require("path");

const upgradeVersion = __filename.replace(/.*upgrade-(.*)\.js$/, "$1");

const DEFAULT_MAPPINGS_JSON = {
  "{vendor}-{uuApp}-{uuSubApp}": {
    useCaseMap: {
      defaultVuc: {
        realization: "index.html",
        httpMethod: "GET",
        type: "VUC"
      }
    }
  }
};

module.exports = async function (config, templateInfo) {
  let { type } = templateInfo;

  if (type === "uu5-app") {
    const PackageManager = require("../tools/package-manager.js");
    let pkgJson = JSON.parse(fs.readFileSync("package.json", "utf-8"));

    if (!("uu5loaderg01" in (pkgJson.dependencies || {}))) {
      console.log(`${upgradeVersion} Installing uu5loaderg01 for use in *.html of UVEs.`);
      await PackageManager.install({ packages: ["uu5loaderg01"], saveExact: false, save: true });
    }

    if ("systemjs" in (pkgJson.dependencies || {})) {
      console.log(`${upgradeVersion} Removing systemjs.`);
      await PackageManager.uninstall({ packages: ["systemjs"], save: true });
    }

    console.log(`${upgradeVersion} Upgrading *.html of UVEs to use uu5loaderg01.`);
    let uveHtmlFiles = getUveHtmlFiles();
    for (let htmlFile of uveHtmlFiles) {
      let content = fs.readFileSync(htmlFile, "utf-8");
      content = content.replace(/~systemjs\/[^"']+/, "~uu5loaderg01");
      content = content.replace(/SystemJS/g, "Uu5Loader");
      fs.writeFileSync(htmlFile, content, "utf-8");
    }
  }
};

function getUveHtmlFiles() {
  let mappingsJson = getMappingsJson();
  let srcPath = "src";

  let resultSet = new Set();
  if (mappingsJson) {
    for (let k in mappingsJson) {
      // iterate over UCs in the mappings.json and pick those which are VUC and end with .html (or .htm)
      let ucMap = (mappingsJson[k] || {})["useCaseMap"] || {};
      Object.keys(ucMap)
        .filter(
          (uc) =>
            (ucMap[uc].type === "VUC" || ucMap[uc].type === "UVE") && (ucMap[uc].realization || "").match(/\.html?$/i)
        )
        .map((uc) => ucMap[uc].realization.replace(/^\/+/, ""))
        .forEach((htmlFile) => resultSet.add(htmlFile));
    }
  }
  let filesInSrc = fs.readdirSync(srcPath, { withFileTypes: true });
  filesInSrc.filter((item) => item.name.match(/\.html?$/) && item.isFile()).forEach((item) => resultSet.add(item.name));
  return [...resultSet].map(it => path.join(srcPath, it));
}

function getMappingsJson() {
  let serverDirName = path.basename(process.cwd()).replace(/-(client|hi)/, "-server"); // "xyz-hi" => "xyz-server"; "xyz-client" => "xyz-server"

  let filePath = `../${serverDirName}/app/config/mappings.json`;
  if (!fs.existsSync(filePath)) filePath = "config/mappings.json";
  if (!fs.existsSync(filePath)) return DEFAULT_MAPPINGS_JSON;
  let mappingsJson = fs.readFileSync(filePath, "utf-8");
  return mappingsJson ? JSON.parse(mappingsJson) : {};
}
