// uu_appg01_core-appclient/src/error/remote-error.js
"use strict";

const { BaseError, Json } = require("uu_appg01_core-utils");

class RemoteError extends BaseError {
  constructor(message = null, response = null) {
    let errorInitialized = false;
    if (typeof message === "string") {
      super(message);
      this.paramMap = {};
      this.dtoOut = {};
      errorInitialized = true;
    } else if (message.uuAppErrorMap) {
      let keys = Object.keys(message.uuAppErrorMap);
      if (keys.length > 0) {
        for (let key of keys) {
          let error = message.uuAppErrorMap[key];
          if (error.type === "error") {
            super(error.message);
            this.code = key;
            if (error.id) {
              this.id = error.id;
            }
            this.paramMap = error.paramMap;
            this.dtoOut = message;
            errorInitialized = true;
            break;
          }
        }
      }
    }
    if (!errorInitialized) {
      super(Json.stringify(message));
      this.paramMap = {};
      this.dtoOut = {};
    }
    if (response) {
      this.response = response;
      this.status = response.status;
    }
  }
}

module.exports = RemoteError;
