// uu_appg01_devkit-common/src/templates-generator/__uu5/decoration-component-image.js
//@@viewOn:imports
import mod from "module";
import { createVisualComponent } from "uu5g05";
import Plus4U5Decoration from "uu_plus4u5decorationg01";
//@@viewOff:imports

const STATICS = {
  //@@viewOn:statics
  uu5Tag: "<%= context.uu5Tag %>",
  sourceUu5Tag: "Plus4U5Decoration.<%= context.decorationComponentName %>",
  //@@viewOff:statics
};

const SIZES = ["l", "m", "s"]; // biggest to lowest
// prettier-ignore
const componentProps = <%= context.decorationComponentProps %>;
const srcFileMapping = <%= context.srcFileMapping %>;
const propTypes = { ...Plus4U5Decoration.<%= context.decorationComponentName %>.propTypes };
for (let k in componentProps) delete propTypes[k]; // some props might be marked as required but we'll have them pre-set

let jsScriptUri = mod?.uri;
if (!jsScriptUri) {
  let lastScript = document.currentScript || Array.prototype.slice.call(document.getElementsByTagName("script"), -1)[0];
  jsScriptUri = lastScript?.src;
}
let libBaseUri = jsScriptUri ? jsScriptUri.toString().replace(/^(.*\/).*/, "$1") : "./";

function normalizeSrc(src) {
  return !src || src.match(/^[a-z]+:/i)
    ? src
    : libBaseUri + "assets/" + src.replace(/^\/+/, "").replace(/\\/g, "/");
}
function resolveMapping(mapping, size, aspectRatio) {
  let sizes = mapping?.[aspectRatio];
  if (sizes && !sizes.includes(size)) {
    let index = ["l", "m", "s"].indexOf(size);
    if (index !== -1) {
      for (let i = index + 1; i < SIZES.length; i++) {
        if (sizes.includes(SIZES[i])) {
          size = SIZES[i];
          break;
        }
      }
    }
  }
  return [size, aspectRatio];
}

const <%= context.uu5Tag %> = createVisualComponent({
  ...STATICS,

  //@@viewOn:propTypes
  propTypes,
  //@@viewOff:propTypes

  //@@viewOn:defaultProps
  defaultProps: {},
  //@@viewOff:defaultProps

  render(directProps) {
    //@@viewOn:private
    let propsNoDefaults = { ...componentProps, ...directProps };
    let { src, size, aspectRatio } = propsNoDefaults;
    if (!aspectRatio) aspectRatio = Plus4U5Decoration.<%= context.decorationComponentName %>.defaultProps?.aspectRatio;
    if (!aspectRatio) aspectRatio = "16x9";
    if (!size) size = Plus4U5Decoration.<%= context.decorationComponentName %>.defaultProps?.size;
    if (!size) size = "m";

    if (!directProps.src && componentProps.src) {
      let [_, name, extension] = componentProps.src.match(/^(.*)(\..*)$/) || [0, componentProps.src, ""];
      let srcBase = normalizeSrc(name);
      [size, aspectRatio] = resolveMapping(srcFileMapping, size, aspectRatio);
      src = `${srcBase}-${size}-${aspectRatio}${extension}`;
    }
    //@@viewOff:private

    //@@viewOn:interface
    //@@viewOff:interface

    //@@viewOn:render
    return <Plus4U5Decoration.<%= context.decorationComponentName %> {...propsNoDefaults} src={src} aspectRatio={aspectRatio} />;
    //@@viewOff:render
  },
});

export { <%= context.uu5Tag %> };
