// uu_appg01_core-authentication/src/client-identity.js
"use strict";

/**
 * The identity of a client application.
 */
class ClientIdentity {
  /**
   * Creates new instance of client application identity.
   * @param {String} uuIdentity Unique identifier of the client application
   * @param {Object} data Additional identity data.
   * @param {String} data.uri Registered URI of client application
   * @param {String} data.product Client application product
   * @param {String} data.vendor Client application vendor
   * @param {String} data.app Client application name
   * @param {String} data.subApp Client application subapp
   * @param {String} data.asid Client application asid
   * @param {String} data.awid Client application awid
   * @param {String} data.type Type of the client identity
   * @param {String} data.authenticationLevelOfAssurance String representation of the means of client identity authentication
   * @param {UuAppKey} data.uuAppKey uuAppKey representing the system identity. Set only if the identity type is system identity.
   *
   * @private
   */
  constructor(uuIdentity, data = {}) {
    this._uuIdentity = uuIdentity;
    if (data) {
      this._uri = data.uri;
      this._product = data.product;
      this._vendor = data.vendor;
      this._app = data.app;
      this._subApp = data.subApp;
      this._asid = data.asid;
      this._awid = data.awid;
      this._type = data.type;
      this._authenticationLevelOfAssurance = data.authenticationLevelOfAssurance;
      this._uuAppKey = data.uuAppKey;
    }
  }

  /**
   * Returns unique identifier of the client application.
   * @returns {String|null} Unique identifier of the client application
   */
  getUuIdentity() {
    return this._uuIdentity;
  }

  /**
   * Returns registered URI of client application.
   * @returns {String|null} Registered URI of client application
   */
  getUri() {
    return this._uri;
  }

  /**
   * Returns client application product.
   * @returns {String|null} Client application product
   */
  getProduct() {
    return this._product;
  }

  /**
   * Returns client application vendor.
   * @returns {String|null} Client application vendor
   */
  getVendor() {
    return this._vendor;
  }

  /**
   * Returns client application name.
   * @returns {String|null} Client application name
   */
  getApp() {
    return this._app;
  }

  /**
   * Returns client application subapp.
   * @returns {String|null} Client application subapp
   */
  getSubApp() {
    return this._subApp;
  }

  /**
   * Returns client application asid.
   * @returns {String|null} Client application asid
   */
  getAsid() {
    return this._asid;
  }

  /**
   * Returns client application awid.
   * @returns {String|null} Client application awid
   */
  getAwid() {
    return this._awid;
  }

  /**
   * Returns type of the client identity.
   * @returns {String|null} Type of the client identity
   */
  getType() {
    return this._type;
  }

  /**
   * Returns the means of authentication (type of credentials) of client identity.
   * @returns {String} String representation of the means of authentication
   */
  getAuthenticationLevelOfAssurance() {
    return this._authenticationLevelOfAssurance;
  }

  /**
   * Returns uuAppKey representation of system identity.
   * @returns {UuAppKey} uuAppKey representation of the system identity
   */
  getUuAppKey() {
    return this._uuAppKey;
  }
}

module.exports = ClientIdentity;
