// uu_appg01_core-appserver/src/index.js
"use strict";

module.exports = {
  get AppServer() {
    return require("./app-server.js");
  },
  get startServer() {
    return require("./app-server.js").startServer;
  },

  get stopServer() {
    return require("./app-server.js").stopServer;
  },
  get disableMiddleware() {
    return require("./app-server.js").disableMiddleware;
  },
  get UseCaseContext() {
    return require("./endpoint/use-case-context.js");
  },
  get UseCaseEnvironment() {
    return require("./endpoint/use-case-environment.js");
  },
  get MappingsLoader() {
    return require("./runner/mappings-loader.js");
  },
  get RouteRegister() {
    return require("./endpoint/route-register.js");
  },
  get ValidationHelper() {
    return require("./utils/validation-helper.js");
  },
  get UveLoader() {
    return require("./utils/uve-loader.js");
  },
  get ErrorConverter() {
    return require("./utils/error-converter");
  },
  get ClientConfigurator() {
    return require("./runner/client-configurator.js");
  },
  get Sys() {
    return require("./sys/sys.js");
  },
  get UseCaseError() {
    return require("./error/use-case-error.js");
  },
  get BadRequest() {
    return require("./error/bad-request.js");
  },
  get NotFound() {
    return require("./error/not-found.js");
  },
  get MethodNotAllowed() {
    return require("./error/method-not-allowed.js");
  },
  get UnsupportedMediaType() {
    return require("./error/unsupported-media-type.js");
  },
  get NotAcceptable() {
    return require("./error/not-acceptable.js");
  },
  get AuthenticationFailure() {
    return require("./error/authentication-failure.js");
  },
  get InvalidCredentials() {
    return require("./error/invalid-credentials.js");
  },
  get UntrustedCredentials() {
    return require("./error/untrusted-credentials.js");
  },
  get AuthorizationFailure() {
    return require("./error/authorization-failure.js");
  },
  get AccessDenied() {
    return require("./error/access-denied.js");
  }
};
