// uu_apprepresentationg01/src/api/errors/sys-product-info-errors.js
const UuApprepresentationError = require("./uu-apprepresentation-error.js");

const setProductInfo = function (prefixCode) {
  const UC_CODE = `${prefixCode}productInfo/set/`;

  return {
    getUnsupportedKeysWarningCode: function () {
      return `${UC_CODE}unsupportedKeys`;
    },

    InvalidDtoInError: class extends UuApprepresentationError {
      constructor() {
        super(...arguments);
        this.message = "DtoIn is not valid.";
        this.code = `${UC_CODE}invalidDtoIn`;
      }
    },

    SysProductInfoDoesNotExistError: class extends UuApprepresentationError {
      constructor() {
        super(...arguments);
        this.message = "SysProductInfo uuObject does not exist in the uuAppObjectStore.";
        this.code = `${UC_CODE}sysProductInfoDoesNotExist`;
      }
    },

    SysProductInfoDaoUpdateFailedError: class extends UuApprepresentationError {
      constructor() {
        super(...arguments);
        this.message = "Failed to update sysProductInfo uuObject.";
        this.code = `${UC_CODE}sysProductInfoDaoUpdateFailed`;
      }
    },

    SysProductInfoDaoCreateFailedError: class extends UuApprepresentationError {
      constructor() {
        super(...arguments);
        this.message = "Failed to create sysProductInfo uuObject.";
        this.code = `${UC_CODE}sysProductInfoDaoCreateFailed`;
      }
    },
  };
};

const getProductInfo = function (prefixCode) {
  const UC_CODE = `${prefixCode}productInfo/get/`;

  return {
    SysProductInfoDoesNotExistError: class extends UuApprepresentationError {
      constructor() {
        super(...arguments);
        this.message = "SysProductInfo uuObject does not exist in the uuAppObjectStore.";
        this.code = `${UC_CODE}sysProductInfoDoesNotExist`;
      }
    },
  };
};

module.exports = function (isAsid) {
  if (isAsid) {
    return {
      SetProductInfo: setProductInfo(UuApprepresentationError.APP_INSTANCE_ERROR_PREFIX),
      GetProductInfo: getProductInfo(UuApprepresentationError.APP_INSTANCE_ERROR_PREFIX),
    };
  } else {
    return {
      SetProductInfo: setProductInfo(UuApprepresentationError.APP_WORKSPACE_ERROR_PREFIX),
      GetProductInfo: getProductInfo(UuApprepresentationError.APP_WORKSPACE_ERROR_PREFIX),
    };
  }
};
