// uu_appg01_workspace/index.js
"use strict";

module.exports = {
  get WorkspaceAuthorizationService() {
    return require("./src/workspace-authorization-service.js");
  },
  get WorkspaceAuthorizationResult() {
    return require("./src/workspace-authorization-result.js");
  },
  get UuAppWorkspaceError() {
    return require("./src/api/errors/workspace-error.js");
  },
  get AppClientTokenService() {
    return require("./src/app-client-token-service.js");
  },
  get AppClientToken() {
    return require("./src/app-client-token.js");
  },
  get AppKey() {
    console.error("DEPRECATED: Component AppKey in uuAppWorkspace library is deprecated, please use UuAppKey component from uuAppServer authentication library.");
    return require("./src/app-key.js");
  },
  get ProfileConfig() {
    return require("./src/helpers/profile-config.js");
  },
  get UuSubAppInstance() {
    return require("./src/api/components/app-instance");
  },
  get UuAppWorkspace() {
    return require("./src/api/components/app-workspace");
  },
  get Profile() {
    return require("./src/api/components/profile");
  },
  get Permission() {
    return require("./src/api/components/permission");
  },
  get UuAppDataStore() {
    return require("./src/api/components/app-data-store");
  },
  get SysUuSubAppInstanceErrors() {
    return require("./src/api/errors/sys-app-instance-errors");
  },
  get SysUuAppWorkspaceErrors() {
    return require("./src/api/errors/sys-app-workspace-errors.js");
  },
  get SysUuAppDataStoreErrors() {
    return require("./src/api/errors/sys-app-data-store-errors.js");
  },
  get SysAsidSysStateErrors() {
    return require("./src/api/errors/sys-asid-sys-state-errors.js");
  },
  get SysAwidSysStateErrors() {
    return require("./src/api/errors/sys-awid-sys-state-errors.js");
  },
  get SysDumpRestoreErrors() {
    return require("./src/api/errors/sys-dump-restore-errors.js");
  },
  get SysAppClientTokenErrors() {
    return require("./src/api/errors/sys-app-client-token-errors.js");
  },
  get SysKeyStoreAbl() {
    return require("./src/abl/sys-key-store-abl.js");
  },
  get SysSecretStoreErrors() {
    return require("./src/api/errors/sys-app-secret-store-errors.js");
  },
  get UuAppSecretStore() {
    return require("./src/api/components/app-secret-store.js");
  },
};
