// uu_appg01_devkit/src/scripts/test.multi.js
const path = require("path");
const { exec, resolveByTemplateType } = require("../tools/helpers.js");

module.exports = class Test {
  constructor(config) {
    this.config = config;
  }

  async process() {
    let modules = this.config.getWorkspaceLibraryList();
    console.log("Testing projects: " + modules.map((it) => it.name).join(", "));

    let hasTestError;
    for (let { name, pkg, type } of modules) {
      if (!pkg.scripts.test || resolveByTemplateType(path.join(__dirname, "test.js"), type) == null) {
        continue;
      }
      try {
        await exec("npm test -- " + this.config.toCommandLineArgs().join(" "), {
          cwd: path.resolve(name),
          stdio: "inherit", // Jest repositions text cursor during tests and has issues with it if stdout/err is piped
          env: {
            ...process.env,
            UU_APPG01_DEVKIT_SPECVC: "", // TODO Encapsulate and do for all xyz.multi.js scripts.
          },
        }); // TODO Escaping of arguments
      } catch (e) {
        // if some tests failed then the whole process exits with non-zero status code, which is ok for
        // us so keep on testing
        hasTestError = true;
      }
    }
    if (hasTestError) process.exit(1);
  }
};
